// tss.js/lib/TpmTypes.js
"use strict";
/*
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See the LICENSE file in the project root for full license information.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var TpmMarshaller_js_1 = require("./TpmMarshaller.js");
/**
* TPM object handle (and related data)
*/
var TPM_HANDLE = (function (_super) {
    __extends(TPM_HANDLE, _super);
    function TPM_HANDLE(handle) {
        if (handle === void 0) { handle = 0; }
        var _this = _super.call(this) || this;
        _this.handle = handle;
        return _this;
    }
    /** TpmMarshaller method */
    TPM_HANDLE.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        return TpmMarshaller_js_1.toTpm(this.handle, buf, 4, pos);
    };
    /** TpmMarshaller method */
    TPM_HANDLE.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 4, pos), this.handle = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM_HANDLE;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPM_HANDLE = TPM_HANDLE;
;
/**
* Table 2:124 - Definition of TPMS_AUTH_COMMAND Structure  (StructuresTable)
*/
var TPMS_AUTH_COMMAND = (function (_super) {
    __extends(TPMS_AUTH_COMMAND, _super);
    function TPMS_AUTH_COMMAND(sessionHandle, nonce, sessionAttributes, hmac) {
        if (sessionHandle === void 0) { sessionHandle = new TPM_HANDLE(0); }
        if (nonce === void 0) { nonce = new Buffer(0); }
        if (sessionAttributes === void 0) { sessionAttributes = 0; }
        if (hmac === void 0) { hmac = new Buffer(0); }
        var _this = _super.call(this) || this;
        _this.sessionHandle = sessionHandle;
        _this.nonce = nonce;
        _this.sessionAttributes = sessionAttributes;
        _this.hmac = hmac;
        return _this;
    }
    /** TpmMarshaller method */
    TPMS_AUTH_COMMAND.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = this.sessionHandle.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.nonce, buf, pos);
        pos = TpmMarshaller_js_1.toTpm(this.sessionAttributes, buf, 1, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.hmac, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_AUTH_COMMAND.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = this.sessionHandle.fromTpm(buf, pos);
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.nonce = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm(buf, 1, pos), this.sessionAttributes = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.hmac = _c[0], pos = _c[1];
        return pos;
        var _a, _b, _c;
    };
    return TPMS_AUTH_COMMAND;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMS_AUTH_COMMAND = TPMS_AUTH_COMMAND;
; // class TPMS_AUTH_COMMAND
// Table 2:125 - Definition of TPMS_AUTH_RESPONSE Structure  (StructuresTable)
var TPMS_AUTH_RESPONSE = (function (_super) {
    __extends(TPMS_AUTH_RESPONSE, _super);
    function TPMS_AUTH_RESPONSE(nonce, sessionAttributes, hmac) {
        if (nonce === void 0) { nonce = new Buffer(0); }
        if (sessionAttributes === void 0) { sessionAttributes = 0; }
        if (hmac === void 0) { hmac = new Buffer(0); }
        var _this = _super.call(this) || this;
        _this.nonce = nonce;
        _this.sessionAttributes = sessionAttributes;
        _this.hmac = hmac;
        return _this;
    }
    /** TpmMarshaller method */
    TPMS_AUTH_RESPONSE.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.nonce, buf, pos);
        pos = TpmMarshaller_js_1.toTpm(this.sessionAttributes, buf, 1, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.hmac, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_AUTH_RESPONSE.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.nonce = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm(buf, 1, pos), this.sessionAttributes = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.hmac = _c[0], pos = _c[1];
        return pos;
        var _a, _b, _c;
    };
    return TPMS_AUTH_RESPONSE;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMS_AUTH_RESPONSE = TPMS_AUTH_RESPONSE;
; // class TPMS_AUTH_RESPONSE
/**
 *  Table 2 is the list of algorithms to which the TCG has assigned an algorithm identifier along with its numeric identifier.
 */
var TPM_ALG_ID;
(function (TPM_ALG_ID) {
    /**
     *  should not occur
     */
    TPM_ALG_ID[TPM_ALG_ID["ERROR"] = 0] = "ERROR";
    /**
     *  the RSA algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["FIRST"] = 1] = "FIRST";
    /**
     *  the RSA algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["RSA"] = 1] = "RSA";
    /**
     *  the SHA1 algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["SHA"] = 4] = "SHA";
    /**
     *  redefinition for documentation consistency
     */
    TPM_ALG_ID[TPM_ALG_ID["SHA1"] = 4] = "SHA1";
    TPM_ALG_ID[TPM_ALG_ID["TDES"] = 3] = "TDES";
    /**
     *  Hash Message Authentication Code (HMAC) algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["HMAC"] = 5] = "HMAC";
    /**
     *  the AES algorithm with various key sizes
     */
    TPM_ALG_ID[TPM_ALG_ID["AES"] = 6] = "AES";
    /**
     *  hash-based mask-generation function
     */
    TPM_ALG_ID[TPM_ALG_ID["MGF1"] = 7] = "MGF1";
    /**
     *  an object type that may use XOR for encryption or an HMAC for signing and may also refer to a data object that is neither signing nor encrypting
     */
    TPM_ALG_ID[TPM_ALG_ID["KEYEDHASH"] = 8] = "KEYEDHASH";
    /**
     *  the XOR encryption algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["XOR"] = 10] = "XOR";
    /**
     *  the SHA 256 algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["SHA256"] = 11] = "SHA256";
    /**
     *  the SHA 384 algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["SHA384"] = 12] = "SHA384";
    /**
     *  the SHA 512 algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["SHA512"] = 13] = "SHA512";
    /**
     *  Null algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["NULL"] = 16] = "NULL";
    /**
     *  SM3 hash algorithm
     */
    TPM_ALG_ID[TPM_ALG_ID["SM3_256"] = 18] = "SM3_256";
    /**
     *  SM4 symmetric block cipher
     */
    TPM_ALG_ID[TPM_ALG_ID["SM4"] = 19] = "SM4";
    /**
     *  a signature algorithm defined in section 8.2 (RSASSA-PKCS1-v1_5)
     */
    TPM_ALG_ID[TPM_ALG_ID["RSASSA"] = 20] = "RSASSA";
    /**
     *  a padding algorithm defined in section 7.2 (RSAES-PKCS1-v1_5)
     */
    TPM_ALG_ID[TPM_ALG_ID["RSAES"] = 21] = "RSAES";
    /**
     *  a signature algorithm defined in section 8.1 (RSASSA-PSS)
     */
    TPM_ALG_ID[TPM_ALG_ID["RSAPSS"] = 22] = "RSAPSS";
    /**
     *  a padding algorithm defined in section 7.1 (RSAES_OAEP)
     */
    TPM_ALG_ID[TPM_ALG_ID["OAEP"] = 23] = "OAEP";
    /**
     *  signature algorithm using elliptic curve cryptography (ECC)
     */
    TPM_ALG_ID[TPM_ALG_ID["ECDSA"] = 24] = "ECDSA";
    /**
     *  secret sharing using ECC Based on context, this can be either One-Pass Diffie-Hellman, C(1, 1, ECC CDH) defined in 6.2.2.2 or Full Unified Model C(2, 2, ECC CDH) defined in 6.1.1.2
     */
    TPM_ALG_ID[TPM_ALG_ID["ECDH"] = 25] = "ECDH";
    /**
     *  elliptic-curve based, anonymous signing scheme
     */
    TPM_ALG_ID[TPM_ALG_ID["ECDAA"] = 26] = "ECDAA";
    /**
     *  SM2  depending on context, either an elliptic-curve based, signature algorithm or a key exchange protocol
     *  NOTE	Type listed as signing but, other uses are allowed according to context.
     */
    TPM_ALG_ID[TPM_ALG_ID["SM2"] = 27] = "SM2";
    /**
     *  elliptic-curve based Schnorr signature
     */
    TPM_ALG_ID[TPM_ALG_ID["ECSCHNORR"] = 28] = "ECSCHNORR";
    /**
     *  two-phase elliptic-curve key exchange  C(2, 2, ECC MQV) section 6.1.1.4
     */
    TPM_ALG_ID[TPM_ALG_ID["ECMQV"] = 29] = "ECMQV";
    /**
     *  concatenation key derivation function (approved alternative 1) section 5.8.1
     */
    TPM_ALG_ID[TPM_ALG_ID["KDF1_SP800_56A"] = 32] = "KDF1_SP800_56A";
    /**
     *  key derivation function KDF2 section 13.2
     */
    TPM_ALG_ID[TPM_ALG_ID["KDF2"] = 33] = "KDF2";
    /**
     *  a key derivation method Section 5.1 KDF in Counter Mode
     */
    TPM_ALG_ID[TPM_ALG_ID["KDF1_SP800_108"] = 34] = "KDF1_SP800_108";
    /**
     *  prime field ECC
     */
    TPM_ALG_ID[TPM_ALG_ID["ECC"] = 35] = "ECC";
    /**
     *  the object type for a symmetric block cipher
     */
    TPM_ALG_ID[TPM_ALG_ID["SYMCIPHER"] = 37] = "SYMCIPHER";
    /**
     *  Camellia is symmetric block cipher. The Camellia algorithm with various key sizes
     */
    TPM_ALG_ID[TPM_ALG_ID["CAMELLIA"] = 38] = "CAMELLIA";
    /**
     *  Counter mode  if implemented, all symmetric block ciphers (S type) implemented shall be capable of using this mode.
     */
    TPM_ALG_ID[TPM_ALG_ID["CTR"] = 64] = "CTR";
    /**
     *  Output Feedback mode  if implemented, all symmetric block ciphers (S type) implemented shall be capable of using this mode.
     */
    TPM_ALG_ID[TPM_ALG_ID["OFB"] = 65] = "OFB";
    /**
     *  Cipher Block Chaining mode  if implemented, all symmetric block ciphers (S type) implemented shall be capable of using this mode.
     */
    TPM_ALG_ID[TPM_ALG_ID["CBC"] = 66] = "CBC";
    /**
     *  Cipher Feedback mode  if implemented, all symmetric block ciphers (S type) implemented shall be capable of using this mode.
     */
    TPM_ALG_ID[TPM_ALG_ID["CFB"] = 67] = "CFB";
    /**
     *  Electronic Codebook mode  if implemented, all symmetric block ciphers (S type) implemented shall be capable of using this mode.
     *  NOTE This mode is not recommended for uses unless the key is frequently rotated such as in video codecs
     */
    TPM_ALG_ID[TPM_ALG_ID["ECB"] = 68] = "ECB";
    TPM_ALG_ID[TPM_ALG_ID["LAST"] = 68] = "LAST";
    /**
     *  Phony alg ID to be used for the first union member with no selector
     */
    TPM_ALG_ID[TPM_ALG_ID["ANY"] = 32767] = "ANY";
    /**
     *  Phony alg ID to be used for the second union member with no selector
     */
    TPM_ALG_ID[TPM_ALG_ID["ANY2"] = 32766] = "ANY2";
})(TPM_ALG_ID = exports.TPM_ALG_ID // UINT16
 || (exports.TPM_ALG_ID // UINT16
 = {}));
; // enum TPM_ALG_ID
/**
 *  Table 3 is the list of identifiers for TCG-registered curve ID values for elliptic curve cryptography.
 */
var TPM_ECC_CURVE;
(function (TPM_ECC_CURVE) {
    TPM_ECC_CURVE[TPM_ECC_CURVE["NONE"] = 0] = "NONE";
    TPM_ECC_CURVE[TPM_ECC_CURVE["NIST_P192"] = 1] = "NIST_P192";
    TPM_ECC_CURVE[TPM_ECC_CURVE["NIST_P224"] = 2] = "NIST_P224";
    TPM_ECC_CURVE[TPM_ECC_CURVE["NIST_P256"] = 3] = "NIST_P256";
    TPM_ECC_CURVE[TPM_ECC_CURVE["NIST_P384"] = 4] = "NIST_P384";
    TPM_ECC_CURVE[TPM_ECC_CURVE["NIST_P521"] = 5] = "NIST_P521";
    /**
     *  curve to support ECDAA
     */
    TPM_ECC_CURVE[TPM_ECC_CURVE["BN_P256"] = 16] = "BN_P256";
    /**
     *  curve to support ECDAA
     */
    TPM_ECC_CURVE[TPM_ECC_CURVE["BN_P638"] = 17] = "BN_P638";
    TPM_ECC_CURVE[TPM_ECC_CURVE["SM2_P256"] = 32] = "SM2_P256";
})(TPM_ECC_CURVE = exports.TPM_ECC_CURVE // UINT16
 || (exports.TPM_ECC_CURVE // UINT16
 = {}));
; // enum TPM_ECC_CURVE
/**
 *  Table 12 lists the command codes and their attributes. The only normative column in this table is the column indicating the command code assigned to a specific command (the "Command Code" column). For all other columns, the command and response tables in TPM 2.0 Part 3 are definitive.
 */
var TPM_CC;
(function (TPM_CC) {
    /**
     *  Compile variable. May decrease based on implementation.
     */
    TPM_CC[TPM_CC["FIRST"] = 287] = "FIRST";
    TPM_CC[TPM_CC["NV_UndefineSpaceSpecial"] = 287] = "NV_UndefineSpaceSpecial";
    TPM_CC[TPM_CC["EvictControl"] = 288] = "EvictControl";
    TPM_CC[TPM_CC["HierarchyControl"] = 289] = "HierarchyControl";
    TPM_CC[TPM_CC["NV_UndefineSpace"] = 290] = "NV_UndefineSpace";
    TPM_CC[TPM_CC["ChangeEPS"] = 292] = "ChangeEPS";
    TPM_CC[TPM_CC["ChangePPS"] = 293] = "ChangePPS";
    TPM_CC[TPM_CC["Clear"] = 294] = "Clear";
    TPM_CC[TPM_CC["ClearControl"] = 295] = "ClearControl";
    TPM_CC[TPM_CC["ClockSet"] = 296] = "ClockSet";
    TPM_CC[TPM_CC["HierarchyChangeAuth"] = 297] = "HierarchyChangeAuth";
    TPM_CC[TPM_CC["NV_DefineSpace"] = 298] = "NV_DefineSpace";
    TPM_CC[TPM_CC["PCR_Allocate"] = 299] = "PCR_Allocate";
    TPM_CC[TPM_CC["PCR_SetAuthPolicy"] = 300] = "PCR_SetAuthPolicy";
    TPM_CC[TPM_CC["PP_Commands"] = 301] = "PP_Commands";
    TPM_CC[TPM_CC["SetPrimaryPolicy"] = 302] = "SetPrimaryPolicy";
    TPM_CC[TPM_CC["FieldUpgradeStart"] = 303] = "FieldUpgradeStart";
    TPM_CC[TPM_CC["ClockRateAdjust"] = 304] = "ClockRateAdjust";
    TPM_CC[TPM_CC["CreatePrimary"] = 305] = "CreatePrimary";
    TPM_CC[TPM_CC["NV_GlobalWriteLock"] = 306] = "NV_GlobalWriteLock";
    TPM_CC[TPM_CC["GetCommandAuditDigest"] = 307] = "GetCommandAuditDigest";
    TPM_CC[TPM_CC["NV_Increment"] = 308] = "NV_Increment";
    TPM_CC[TPM_CC["NV_SetBits"] = 309] = "NV_SetBits";
    TPM_CC[TPM_CC["NV_Extend"] = 310] = "NV_Extend";
    TPM_CC[TPM_CC["NV_Write"] = 311] = "NV_Write";
    TPM_CC[TPM_CC["NV_WriteLock"] = 312] = "NV_WriteLock";
    TPM_CC[TPM_CC["DictionaryAttackLockReset"] = 313] = "DictionaryAttackLockReset";
    TPM_CC[TPM_CC["DictionaryAttackParameters"] = 314] = "DictionaryAttackParameters";
    TPM_CC[TPM_CC["NV_ChangeAuth"] = 315] = "NV_ChangeAuth";
    /**
     *  PCR
     */
    TPM_CC[TPM_CC["PCR_Event"] = 316] = "PCR_Event";
    /**
     *  PCR
     */
    TPM_CC[TPM_CC["PCR_Reset"] = 317] = "PCR_Reset";
    TPM_CC[TPM_CC["SequenceComplete"] = 318] = "SequenceComplete";
    TPM_CC[TPM_CC["SetAlgorithmSet"] = 319] = "SetAlgorithmSet";
    TPM_CC[TPM_CC["SetCommandCodeAuditStatus"] = 320] = "SetCommandCodeAuditStatus";
    TPM_CC[TPM_CC["FieldUpgradeData"] = 321] = "FieldUpgradeData";
    TPM_CC[TPM_CC["IncrementalSelfTest"] = 322] = "IncrementalSelfTest";
    TPM_CC[TPM_CC["SelfTest"] = 323] = "SelfTest";
    TPM_CC[TPM_CC["Startup"] = 324] = "Startup";
    TPM_CC[TPM_CC["Shutdown"] = 325] = "Shutdown";
    TPM_CC[TPM_CC["StirRandom"] = 326] = "StirRandom";
    TPM_CC[TPM_CC["ActivateCredential"] = 327] = "ActivateCredential";
    TPM_CC[TPM_CC["Certify"] = 328] = "Certify";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyNV"] = 329] = "PolicyNV";
    TPM_CC[TPM_CC["CertifyCreation"] = 330] = "CertifyCreation";
    TPM_CC[TPM_CC["Duplicate"] = 331] = "Duplicate";
    TPM_CC[TPM_CC["GetTime"] = 332] = "GetTime";
    TPM_CC[TPM_CC["GetSessionAuditDigest"] = 333] = "GetSessionAuditDigest";
    TPM_CC[TPM_CC["NV_Read"] = 334] = "NV_Read";
    TPM_CC[TPM_CC["NV_ReadLock"] = 335] = "NV_ReadLock";
    TPM_CC[TPM_CC["ObjectChangeAuth"] = 336] = "ObjectChangeAuth";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicySecret"] = 337] = "PolicySecret";
    TPM_CC[TPM_CC["Rewrap"] = 338] = "Rewrap";
    TPM_CC[TPM_CC["Create"] = 339] = "Create";
    TPM_CC[TPM_CC["ECDH_ZGen"] = 340] = "ECDH_ZGen";
    TPM_CC[TPM_CC["HMAC"] = 341] = "HMAC";
    TPM_CC[TPM_CC["Import"] = 342] = "Import";
    TPM_CC[TPM_CC["Load"] = 343] = "Load";
    TPM_CC[TPM_CC["Quote"] = 344] = "Quote";
    TPM_CC[TPM_CC["RSA_Decrypt"] = 345] = "RSA_Decrypt";
    TPM_CC[TPM_CC["HMAC_Start"] = 347] = "HMAC_Start";
    TPM_CC[TPM_CC["SequenceUpdate"] = 348] = "SequenceUpdate";
    TPM_CC[TPM_CC["Sign"] = 349] = "Sign";
    TPM_CC[TPM_CC["Unseal"] = 350] = "Unseal";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicySigned"] = 352] = "PolicySigned";
    /**
     *  Context
     */
    TPM_CC[TPM_CC["ContextLoad"] = 353] = "ContextLoad";
    /**
     *  Context
     */
    TPM_CC[TPM_CC["ContextSave"] = 354] = "ContextSave";
    TPM_CC[TPM_CC["ECDH_KeyGen"] = 355] = "ECDH_KeyGen";
    TPM_CC[TPM_CC["EncryptDecrypt"] = 356] = "EncryptDecrypt";
    /**
     *  Context
     */
    TPM_CC[TPM_CC["FlushContext"] = 357] = "FlushContext";
    TPM_CC[TPM_CC["LoadExternal"] = 359] = "LoadExternal";
    TPM_CC[TPM_CC["MakeCredential"] = 360] = "MakeCredential";
    /**
     *  NV
     */
    TPM_CC[TPM_CC["NV_ReadPublic"] = 361] = "NV_ReadPublic";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyAuthorize"] = 362] = "PolicyAuthorize";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyAuthValue"] = 363] = "PolicyAuthValue";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyCommandCode"] = 364] = "PolicyCommandCode";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyCounterTimer"] = 365] = "PolicyCounterTimer";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyCpHash"] = 366] = "PolicyCpHash";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyLocality"] = 367] = "PolicyLocality";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyNameHash"] = 368] = "PolicyNameHash";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyOR"] = 369] = "PolicyOR";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyTicket"] = 370] = "PolicyTicket";
    TPM_CC[TPM_CC["ReadPublic"] = 371] = "ReadPublic";
    TPM_CC[TPM_CC["RSA_Encrypt"] = 372] = "RSA_Encrypt";
    TPM_CC[TPM_CC["StartAuthSession"] = 374] = "StartAuthSession";
    TPM_CC[TPM_CC["VerifySignature"] = 375] = "VerifySignature";
    TPM_CC[TPM_CC["ECC_Parameters"] = 376] = "ECC_Parameters";
    TPM_CC[TPM_CC["FirmwareRead"] = 377] = "FirmwareRead";
    TPM_CC[TPM_CC["GetCapability"] = 378] = "GetCapability";
    TPM_CC[TPM_CC["GetRandom"] = 379] = "GetRandom";
    TPM_CC[TPM_CC["GetTestResult"] = 380] = "GetTestResult";
    TPM_CC[TPM_CC["Hash"] = 381] = "Hash";
    /**
     *  PCR
     */
    TPM_CC[TPM_CC["PCR_Read"] = 382] = "PCR_Read";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyPCR"] = 383] = "PolicyPCR";
    TPM_CC[TPM_CC["PolicyRestart"] = 384] = "PolicyRestart";
    TPM_CC[TPM_CC["ReadClock"] = 385] = "ReadClock";
    TPM_CC[TPM_CC["PCR_Extend"] = 386] = "PCR_Extend";
    TPM_CC[TPM_CC["PCR_SetAuthValue"] = 387] = "PCR_SetAuthValue";
    TPM_CC[TPM_CC["NV_Certify"] = 388] = "NV_Certify";
    TPM_CC[TPM_CC["EventSequenceComplete"] = 389] = "EventSequenceComplete";
    TPM_CC[TPM_CC["HashSequenceStart"] = 390] = "HashSequenceStart";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyPhysicalPresence"] = 391] = "PolicyPhysicalPresence";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyDuplicationSelect"] = 392] = "PolicyDuplicationSelect";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyGetDigest"] = 393] = "PolicyGetDigest";
    TPM_CC[TPM_CC["TestParms"] = 394] = "TestParms";
    TPM_CC[TPM_CC["Commit"] = 395] = "Commit";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyPassword"] = 396] = "PolicyPassword";
    TPM_CC[TPM_CC["ZGen_2Phase"] = 397] = "ZGen_2Phase";
    TPM_CC[TPM_CC["EC_Ephemeral"] = 398] = "EC_Ephemeral";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyNvWritten"] = 399] = "PolicyNvWritten";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyTemplate"] = 400] = "PolicyTemplate";
    TPM_CC[TPM_CC["CreateLoaded"] = 401] = "CreateLoaded";
    /**
     *  Policy
     */
    TPM_CC[TPM_CC["PolicyAuthorizeNV"] = 402] = "PolicyAuthorizeNV";
    TPM_CC[TPM_CC["EncryptDecrypt2"] = 403] = "EncryptDecrypt2";
    /**
     *  Compile variable. May increase based on implementation.
     */
    TPM_CC[TPM_CC["LAST"] = 403] = "LAST";
    TPM_CC[TPM_CC["CC_VEND"] = 536870912] = "CC_VEND";
    /**
     *  Used for testing of command dispatch
     */
    TPM_CC[TPM_CC["Vendor_TCG_Test"] = 536870912] = "Vendor_TCG_Test";
})(TPM_CC = exports.TPM_CC // UINT32
 || (exports.TPM_CC // UINT32
 = {}));
; // enum TPM_CC
/**
 *  In general, response codes defined in TPM 2.0 Part 2 will be unmarshaling errors and will have the F (format) bit SET. Codes that are unique to TPM 2.0 Part 3 will have the F bit CLEAR but the V (version) attribute will be SET to indicate that it is a TPM 2.0 response code. See Response Code Details in TPM 2.0 Part 1.
 */
var TPM_RC;
(function (TPM_RC) {
    TPM_RC[TPM_RC["SUCCESS"] = 0] = "SUCCESS";
    /**
     *  defined for compatibility with TPM 1.2
     */
    TPM_RC[TPM_RC["BAD_TAG"] = 30] = "BAD_TAG";
    /**
     *  set for all format 0 response codes
     */
    TPM_RC[TPM_RC["RC_VER1"] = 256] = "RC_VER1";
    /**
     *  TPM not initialized by TPM2_Startup or already initialized
     */
    TPM_RC[TPM_RC["INITIALIZE"] = 256] = "INITIALIZE";
    /**
     *  commands not being accepted because of a TPM failure
     *  NOTE	This may be returned by TPM2_GetTestResult() as the testResult parameter.
     */
    TPM_RC[TPM_RC["FAILURE"] = 257] = "FAILURE";
    /**
     *  improper use of a sequence handle
     */
    TPM_RC[TPM_RC["SEQUENCE"] = 259] = "SEQUENCE";
    /**
     *  not currently used
     */
    TPM_RC[TPM_RC["PRIVATE"] = 267] = "PRIVATE";
    /**
     *  not currently used
     */
    TPM_RC[TPM_RC["HMAC"] = 281] = "HMAC";
    /**
     *  the command is disabled
     */
    TPM_RC[TPM_RC["DISABLED"] = 288] = "DISABLED";
    /**
     *  command failed because audit sequence required exclusivity
     */
    TPM_RC[TPM_RC["EXCLUSIVE"] = 289] = "EXCLUSIVE";
    /**
     *  authorization handle is not correct for command
     */
    TPM_RC[TPM_RC["AUTH_TYPE"] = 292] = "AUTH_TYPE";
    /**
     *  command requires an authorization session for handle and it is not present.
     */
    TPM_RC[TPM_RC["AUTH_MISSING"] = 293] = "AUTH_MISSING";
    /**
     *  policy failure in math operation or an invalid authPolicy value
     */
    TPM_RC[TPM_RC["POLICY"] = 294] = "POLICY";
    /**
     *  PCR check fail
     */
    TPM_RC[TPM_RC["PCR"] = 295] = "PCR";
    /**
     *  PCR have changed since checked.
     */
    TPM_RC[TPM_RC["PCR_CHANGED"] = 296] = "PCR_CHANGED";
    /**
     *  for all commands other than TPM2_FieldUpgradeData(), this code indicates that the TPM is in field upgrade mode; for TPM2_FieldUpgradeData(), this code indicates that the TPM is not in field upgrade mode
     */
    TPM_RC[TPM_RC["UPGRADE"] = 301] = "UPGRADE";
    /**
     *  context ID counter is at maximum.
     */
    TPM_RC[TPM_RC["TOO_MANY_CONTEXTS"] = 302] = "TOO_MANY_CONTEXTS";
    /**
     *  authValue or authPolicy is not available for selected entity.
     */
    TPM_RC[TPM_RC["AUTH_UNAVAILABLE"] = 303] = "AUTH_UNAVAILABLE";
    /**
     *  a _TPM_Init and Startup(CLEAR) is required before the TPM can resume operation.
     */
    TPM_RC[TPM_RC["REBOOT"] = 304] = "REBOOT";
    /**
     *  the protection algorithms (hash and symmetric) are not reasonably balanced. The digest size of the hash must be larger than the key size of the symmetric algorithm.
     */
    TPM_RC[TPM_RC["UNBALANCED"] = 305] = "UNBALANCED";
    /**
     *  command commandSize value is inconsistent with contents of the command buffer; either the size is not the same as the octets loaded by the hardware interface layer or the value is not large enough to hold a command header
     */
    TPM_RC[TPM_RC["COMMAND_SIZE"] = 322] = "COMMAND_SIZE";
    /**
     *  command code not supported
     */
    TPM_RC[TPM_RC["COMMAND_CODE"] = 323] = "COMMAND_CODE";
    /**
     *  the value of authorizationSize is out of range or the number of octets in the Authorization Area is greater than required
     */
    TPM_RC[TPM_RC["AUTHSIZE"] = 324] = "AUTHSIZE";
    /**
     *  use of an authorization session with a context command or another command that cannot have an authorization session.
     */
    TPM_RC[TPM_RC["AUTH_CONTEXT"] = 325] = "AUTH_CONTEXT";
    /**
     *  NV offset+size is out of range.
     */
    TPM_RC[TPM_RC["NV_RANGE"] = 326] = "NV_RANGE";
    /**
     *  Requested allocation size is larger than allowed.
     */
    TPM_RC[TPM_RC["NV_SIZE"] = 327] = "NV_SIZE";
    /**
     *  NV access locked.
     */
    TPM_RC[TPM_RC["NV_LOCKED"] = 328] = "NV_LOCKED";
    /**
     *  NV access authorization fails in command actions (this failure does not affect lockout.action)
     */
    TPM_RC[TPM_RC["NV_AUTHORIZATION"] = 329] = "NV_AUTHORIZATION";
    /**
     *  an NV Index is used before being initialized or the state saved by TPM2_Shutdown(STATE) could not be restored
     */
    TPM_RC[TPM_RC["NV_UNINITIALIZED"] = 330] = "NV_UNINITIALIZED";
    /**
     *  insufficient space for NV allocation
     */
    TPM_RC[TPM_RC["NV_SPACE"] = 331] = "NV_SPACE";
    /**
     *  NV Index or persistent object already defined
     */
    TPM_RC[TPM_RC["NV_DEFINED"] = 332] = "NV_DEFINED";
    /**
     *  context in TPM2_ContextLoad() is not valid
     */
    TPM_RC[TPM_RC["BAD_CONTEXT"] = 336] = "BAD_CONTEXT";
    /**
     *  cpHash value already set or not correct for use
     */
    TPM_RC[TPM_RC["CPHASH"] = 337] = "CPHASH";
    /**
     *  handle for parent is not a valid parent
     */
    TPM_RC[TPM_RC["PARENT"] = 338] = "PARENT";
    /**
     *  some function needs testing.
     */
    TPM_RC[TPM_RC["NEEDS_TEST"] = 339] = "NEEDS_TEST";
    /**
     *  returned when an internal function cannot process a request due to an unspecified problem. This code is usually related to invalid parameters that are not properly filtered by the input unmarshaling code.
     */
    TPM_RC[TPM_RC["NO_RESULT"] = 340] = "NO_RESULT";
    /**
     *  the sensitive area did not unmarshal correctly after decryption  this code is used in lieu of the other unmarshaling errors so that an attacker cannot determine where the unmarshaling error occurred
     */
    TPM_RC[TPM_RC["SENSITIVE"] = 341] = "SENSITIVE";
    /**
     *  largest version 1 code that is not a warning
     */
    TPM_RC[TPM_RC["RC_MAX_FM0"] = 383] = "RC_MAX_FM0";
    /**
     *  This bit is SET in all format 1 response codes
     *  The codes in this group may have a value added to them to indicate the handle, session, or parameter to which they apply.
     */
    TPM_RC[TPM_RC["RC_FMT1"] = 128] = "RC_FMT1";
    /**
     *  asymmetric algorithm not supported or not correct
     */
    TPM_RC[TPM_RC["ASYMMETRIC"] = 129] = "ASYMMETRIC";
    /**
     *  inconsistent attributes
     */
    TPM_RC[TPM_RC["ATTRIBUTES"] = 130] = "ATTRIBUTES";
    /**
     *  hash algorithm not supported or not appropriate
     */
    TPM_RC[TPM_RC["HASH"] = 131] = "HASH";
    /**
     *  value is out of range or is not correct for the context
     */
    TPM_RC[TPM_RC["VALUE"] = 132] = "VALUE";
    /**
     *  hierarchy is not enabled or is not correct for the use
     */
    TPM_RC[TPM_RC["HIERARCHY"] = 133] = "HIERARCHY";
    /**
     *  key size is not supported
     */
    TPM_RC[TPM_RC["KEY_SIZE"] = 135] = "KEY_SIZE";
    /**
     *  mask generation function not supported
     */
    TPM_RC[TPM_RC["MGF"] = 136] = "MGF";
    /**
     *  mode of operation not supported
     */
    TPM_RC[TPM_RC["MODE"] = 137] = "MODE";
    /**
     *  the type of the value is not appropriate for the use
     */
    TPM_RC[TPM_RC["TYPE"] = 138] = "TYPE";
    /**
     *  the handle is not correct for the use
     */
    TPM_RC[TPM_RC["HANDLE"] = 139] = "HANDLE";
    /**
     *  unsupported key derivation function or function not appropriate for use
     */
    TPM_RC[TPM_RC["KDF"] = 140] = "KDF";
    /**
     *  value was out of allowed range.
     */
    TPM_RC[TPM_RC["RANGE"] = 141] = "RANGE";
    /**
     *  the authorization HMAC check failed and DA counter incremented
     */
    TPM_RC[TPM_RC["AUTH_FAIL"] = 142] = "AUTH_FAIL";
    /**
     *  invalid nonce size or nonce value mismatch
     */
    TPM_RC[TPM_RC["NONCE"] = 143] = "NONCE";
    /**
     *  authorization requires assertion of PP
     */
    TPM_RC[TPM_RC["PP"] = 144] = "PP";
    /**
     *  unsupported or incompatible scheme
     */
    TPM_RC[TPM_RC["SCHEME"] = 146] = "SCHEME";
    /**
     *  structure is the wrong size
     */
    TPM_RC[TPM_RC["SIZE"] = 149] = "SIZE";
    /**
     *  unsupported symmetric algorithm or key size, or not appropriate for instance
     */
    TPM_RC[TPM_RC["SYMMETRIC"] = 150] = "SYMMETRIC";
    /**
     *  incorrect structure tag
     */
    TPM_RC[TPM_RC["TAG"] = 151] = "TAG";
    /**
     *  union selector is incorrect
     */
    TPM_RC[TPM_RC["SELECTOR"] = 152] = "SELECTOR";
    /**
     *  the TPM was unable to unmarshal a value because there were not enough octets in the input buffer
     */
    TPM_RC[TPM_RC["INSUFFICIENT"] = 154] = "INSUFFICIENT";
    /**
     *  the signature is not valid
     */
    TPM_RC[TPM_RC["SIGNATURE"] = 155] = "SIGNATURE";
    /**
     *  key fields are not compatible with the selected use
     */
    TPM_RC[TPM_RC["KEY"] = 156] = "KEY";
    /**
     *  a policy check failed
     */
    TPM_RC[TPM_RC["POLICY_FAIL"] = 157] = "POLICY_FAIL";
    /**
     *  integrity check failed
     */
    TPM_RC[TPM_RC["INTEGRITY"] = 159] = "INTEGRITY";
    /**
     *  invalid ticket
     */
    TPM_RC[TPM_RC["TICKET"] = 160] = "TICKET";
    /**
     *  authorization failure without DA implications
     */
    TPM_RC[TPM_RC["BAD_AUTH"] = 162] = "BAD_AUTH";
    /**
     *  the policy has expired
     */
    TPM_RC[TPM_RC["EXPIRED"] = 163] = "EXPIRED";
    /**
     *  the commandCode in the policy is not the commandCode of the command or the command code in a policy command references a command that is not implemented
     */
    TPM_RC[TPM_RC["POLICY_CC"] = 164] = "POLICY_CC";
    /**
     *  public and sensitive portions of an object are not cryptographically bound
     */
    TPM_RC[TPM_RC["BINDING"] = 165] = "BINDING";
    /**
     *  curve not supported
     */
    TPM_RC[TPM_RC["CURVE"] = 166] = "CURVE";
    /**
     *  point is not on the required curve.
     */
    TPM_RC[TPM_RC["ECC_POINT"] = 167] = "ECC_POINT";
    /**
     *  set for warning response codes
     */
    TPM_RC[TPM_RC["RC_WARN"] = 2304] = "RC_WARN";
    /**
     *  gap for context ID is too large
     */
    TPM_RC[TPM_RC["CONTEXT_GAP"] = 2305] = "CONTEXT_GAP";
    /**
     *  out of memory for object contexts
     */
    TPM_RC[TPM_RC["OBJECT_MEMORY"] = 2306] = "OBJECT_MEMORY";
    /**
     *  out of memory for session contexts
     */
    TPM_RC[TPM_RC["SESSION_MEMORY"] = 2307] = "SESSION_MEMORY";
    /**
     *  out of shared object/session memory or need space for internal operations
     */
    TPM_RC[TPM_RC["MEMORY"] = 2308] = "MEMORY";
    /**
     *  out of session handles  a session must be flushed before a new session may be created
     */
    TPM_RC[TPM_RC["SESSION_HANDLES"] = 2309] = "SESSION_HANDLES";
    /**
     *  out of object handles  the handle space for objects is depleted and a reboot is required
     *  NOTE 1	This cannot occur on the reference implementation.
     *  NOTE 2	There is no reason why an implementation would implement a design that would deplete handle space. Platform specifications are encouraged to forbid it.
     */
    TPM_RC[TPM_RC["OBJECT_HANDLES"] = 2310] = "OBJECT_HANDLES";
    /**
     *  bad locality
     */
    TPM_RC[TPM_RC["LOCALITY"] = 2311] = "LOCALITY";
    /**
     *  the TPM has suspended operation on the command; forward progress was made and the command may be retried
     *  See TPM 2.0 Part 1, Multi-tasking.
     *  NOTE	This cannot occur on the reference implementation.
     */
    TPM_RC[TPM_RC["YIELDED"] = 2312] = "YIELDED";
    /**
     *  the command was canceled
     */
    TPM_RC[TPM_RC["CANCELED"] = 2313] = "CANCELED";
    /**
     *  TPM is performing self-tests
     */
    TPM_RC[TPM_RC["TESTING"] = 2314] = "TESTING";
    /**
     *  the 1st handle in the handle area references a transient object or session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_H0"] = 2320] = "REFERENCE_H0";
    /**
     *  the 2nd handle in the handle area references a transient object or session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_H1"] = 2321] = "REFERENCE_H1";
    /**
     *  the 3rd handle in the handle area references a transient object or session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_H2"] = 2322] = "REFERENCE_H2";
    /**
     *  the 4th handle in the handle area references a transient object or session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_H3"] = 2323] = "REFERENCE_H3";
    /**
     *  the 5th handle in the handle area references a transient object or session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_H4"] = 2324] = "REFERENCE_H4";
    /**
     *  the 6th handle in the handle area references a transient object or session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_H5"] = 2325] = "REFERENCE_H5";
    /**
     *  the 7th handle in the handle area references a transient object or session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_H6"] = 2326] = "REFERENCE_H6";
    /**
     *  the 1st authorization session handle references a session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_S0"] = 2328] = "REFERENCE_S0";
    /**
     *  the 2nd authorization session handle references a session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_S1"] = 2329] = "REFERENCE_S1";
    /**
     *  the 3rd authorization session handle references a session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_S2"] = 2330] = "REFERENCE_S2";
    /**
     *  the 4th authorization session handle references a session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_S3"] = 2331] = "REFERENCE_S3";
    /**
     *  the 5th session handle references a session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_S4"] = 2332] = "REFERENCE_S4";
    /**
     *  the 6th session handle references a session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_S5"] = 2333] = "REFERENCE_S5";
    /**
     *  the 7th authorization session handle references a session that is not loaded
     */
    TPM_RC[TPM_RC["REFERENCE_S6"] = 2334] = "REFERENCE_S6";
    /**
     *  the TPM is rate-limiting accesses to prevent wearout of NV
     */
    TPM_RC[TPM_RC["NV_RATE"] = 2336] = "NV_RATE";
    /**
     *  authorizations for objects subject to DA protection are not allowed at this time because the TPM is in DA lockout mode
     */
    TPM_RC[TPM_RC["LOCKOUT"] = 2337] = "LOCKOUT";
    /**
     *  the TPM was not able to start the command
     */
    TPM_RC[TPM_RC["RETRY"] = 2338] = "RETRY";
    /**
     *  the command may require writing of NV and NV is not current accessible
     */
    TPM_RC[TPM_RC["NV_UNAVAILABLE"] = 2339] = "NV_UNAVAILABLE";
    /**
     *  this value is reserved and shall not be returned by the TPM
     */
    TPM_RC[TPM_RC["NOT_USED"] = 2431] = "NOT_USED";
})(TPM_RC = exports.TPM_RC // UINT32
 || (exports.TPM_RC // UINT32
 = {}));
; // enum TPM_RC
/**
 *  Structure tags are used to disambiguate structures. They are 16-bit values with the most significant bit SET so that they do not overlap TPM_ALG_ID values. A single exception is made for the value associated with TPM_ST_RSP_COMMAND (0x00C4), which has the same value as the TPM_TAG_RSP_COMMAND tag from earlier versions of this specification. This value is used when the TPM is compatible with a previous TPM specification and the TPM cannot determine which family of response code to return because the command tag is not valid.
 */
var TPM_ST;
(function (TPM_ST) {
    /**
     *  tag value for a response; used when there is an error in the tag. This is also the value returned from a TPM 1.2 when an error occurs. This value is used in this specification because an error in the command tag may prevent determination of the family. When this tag is used in the response, the response code will be TPM_RC_BAD_TAG (0 1E16), which has the same numeric value as the TPM 1.2 response code for TPM_BADTAG.
     *  NOTE	In a previously published version of this specification, TPM_RC_BAD_TAG was incorrectly assigned a value of 0x030 instead of 30 (0x01e). Some implementations my return the old value instead of the new value.
     */
    TPM_ST[TPM_ST["RSP_COMMAND"] = 196] = "RSP_COMMAND";
    /**
     *  no structure type specified
     */
    TPM_ST[TPM_ST["_NULL"] = 32768] = "_NULL";
    /**
     *  tag value for a command/response for a command defined in this specification; indicating that the command/response has no attached sessions and no authorizationSize/parameterSize value is present
     *  If the responseCode from the TPM is not TPM_RC_SUCCESS, then the response tag shall have this value.
     */
    TPM_ST[TPM_ST["NO_SESSIONS"] = 32769] = "NO_SESSIONS";
    /**
     *  tag value for a command/response for a command defined in this specification; indicating that the command/response has one or more attached sessions and the authorizationSize/parameterSize field is present
     */
    TPM_ST[TPM_ST["SESSIONS"] = 32770] = "SESSIONS";
    /**
     *  tag for an attestation structure
     */
    TPM_ST[TPM_ST["ATTEST_NV"] = 32788] = "ATTEST_NV";
    /**
     *  tag for an attestation structure
     */
    TPM_ST[TPM_ST["ATTEST_COMMAND_AUDIT"] = 32789] = "ATTEST_COMMAND_AUDIT";
    /**
     *  tag for an attestation structure
     */
    TPM_ST[TPM_ST["ATTEST_SESSION_AUDIT"] = 32790] = "ATTEST_SESSION_AUDIT";
    /**
     *  tag for an attestation structure
     */
    TPM_ST[TPM_ST["ATTEST_CERTIFY"] = 32791] = "ATTEST_CERTIFY";
    /**
     *  tag for an attestation structure
     */
    TPM_ST[TPM_ST["ATTEST_QUOTE"] = 32792] = "ATTEST_QUOTE";
    /**
     *  tag for an attestation structure
     */
    TPM_ST[TPM_ST["ATTEST_TIME"] = 32793] = "ATTEST_TIME";
    /**
     *  tag for an attestation structure
     */
    TPM_ST[TPM_ST["ATTEST_CREATION"] = 32794] = "ATTEST_CREATION";
    /**
     *  tag for a ticket type
     */
    TPM_ST[TPM_ST["CREATION"] = 32801] = "CREATION";
    /**
     *  tag for a ticket type
     */
    TPM_ST[TPM_ST["VERIFIED"] = 32802] = "VERIFIED";
    /**
     *  tag for a ticket type
     */
    TPM_ST[TPM_ST["AUTH_SECRET"] = 32803] = "AUTH_SECRET";
    /**
     *  tag for a ticket type
     */
    TPM_ST[TPM_ST["HASHCHECK"] = 32804] = "HASHCHECK";
    /**
     *  tag for a ticket type
     */
    TPM_ST[TPM_ST["AUTH_SIGNED"] = 32805] = "AUTH_SIGNED";
    /**
     *  tag for a structure describing a Field Upgrade Policy
     */
    TPM_ST[TPM_ST["FU_MANIFEST"] = 32809] = "FU_MANIFEST";
})(TPM_ST = exports.TPM_ST // UINT16
 || (exports.TPM_ST // UINT16
 = {}));
; // enum TPM_ST
/**
 *  These values are used in TPM2_Startup() to indicate the shutdown and startup mode. The defined startup sequences are:
 */
var TPM_SU;
(function (TPM_SU) {
    /**
     *  on TPM2_Shutdown(), indicates that the TPM should prepare for loss of power and save state required for an orderly startup (TPM Reset).
     *  on TPM2_Startup(), indicates that the TPM should perform TPM Reset or TPM Restart
     */
    TPM_SU[TPM_SU["CLEAR"] = 0] = "CLEAR";
    /**
     *  on TPM2_Shutdown(), indicates that the TPM should prepare for loss of power and save state required for an orderly startup (TPM Restart or TPM Resume)
     *  on TPM2_Startup(), indicates that the TPM should restore the state saved by TPM2_Shutdown(TPM_SU_STATE)
     */
    TPM_SU[TPM_SU["STATE"] = 1] = "STATE";
})(TPM_SU = exports.TPM_SU // UINT16
 || (exports.TPM_SU // UINT16
 = {}));
; // enum TPM_SU
/**
 *  This type is used in TPM2_StartAuthSession() to indicate the type of the session to be created.
 */
var TPM_SE;
(function (TPM_SE) {
    TPM_SE[TPM_SE["HMAC"] = 0] = "HMAC";
    TPM_SE[TPM_SE["POLICY"] = 1] = "POLICY";
    /**
     *  The policy session is being used to compute the policyHash and not for command authorization.
     *  This setting modifies some policy commands and prevents session from being used to authorize a command.
     */
    TPM_SE[TPM_SE["TRIAL"] = 3] = "TRIAL";
})(TPM_SE = exports.TPM_SE // BYTE
 || (exports.TPM_SE // BYTE
 = {}));
; // enum TPM_SE
/**
 *  The TPM_CAP values are used in TPM2_GetCapability() to select the type of the value to be returned. The format of the response varies according to the type of the value.
 */
var TPM_CAP;
(function (TPM_CAP) {
    TPM_CAP[TPM_CAP["FIRST"] = 0] = "FIRST";
    /**
     *  TPML_ALG_PROPERTY
     */
    TPM_CAP[TPM_CAP["ALGS"] = 0] = "ALGS";
    /**
     *  TPML_HANDLE
     */
    TPM_CAP[TPM_CAP["HANDLES"] = 1] = "HANDLES";
    /**
     *  TPML_CCA
     */
    TPM_CAP[TPM_CAP["COMMANDS"] = 2] = "COMMANDS";
    /**
     *  TPML_CC
     */
    TPM_CAP[TPM_CAP["PP_COMMANDS"] = 3] = "PP_COMMANDS";
    /**
     *  TPML_CC
     */
    TPM_CAP[TPM_CAP["AUDIT_COMMANDS"] = 4] = "AUDIT_COMMANDS";
    /**
     *  TPML_PCR_SELECTION
     */
    TPM_CAP[TPM_CAP["PCRS"] = 5] = "PCRS";
    /**
     *  TPML_TAGGED_TPM_PROPERTY
     */
    TPM_CAP[TPM_CAP["TPM_PROPERTIES"] = 6] = "TPM_PROPERTIES";
    /**
     *  TPML_TAGGED_PCR_PROPERTY
     */
    TPM_CAP[TPM_CAP["PCR_PROPERTIES"] = 7] = "PCR_PROPERTIES";
    /**
     *  TPML_ECC_CURVE
     */
    TPM_CAP[TPM_CAP["ECC_CURVES"] = 8] = "ECC_CURVES";
    /**
     *  TPML_TAGGED_POLICY
     */
    TPM_CAP[TPM_CAP["AUTH_POLICIES"] = 9] = "AUTH_POLICIES";
    TPM_CAP[TPM_CAP["LAST"] = 9] = "LAST";
    /**
     *  manufacturer-specific values
     */
    TPM_CAP[TPM_CAP["VENDOR_PROPERTY"] = 256] = "VENDOR_PROPERTY";
})(TPM_CAP = exports.TPM_CAP // UINT32
 || (exports.TPM_CAP // UINT32
 = {}));
; // enum TPM_CAP
/**
 *  The TPM_PT constants are used in TPM2_GetCapability(capability = TPM_CAP_TPM_PROPERTIES) to indicate the property being selected or returned.
 */
var TPM_PT;
(function (TPM_PT) {
    /**
     *  indicates no property type
     */
    TPM_PT[TPM_PT["NONE"] = 0] = "NONE";
    /**
     *  The number of properties in each group.
     *  NOTE The first group with any properties is group 1 (PT_GROUP * 1). Group 0 is reserved.
     */
    TPM_PT[TPM_PT["PT_GROUP"] = 256] = "PT_GROUP";
    /**
     *  the group of fixed properties returned as TPMS_TAGGED_PROPERTY
     *  The values in this group are only changed due to a firmware change in the TPM.
     */
    TPM_PT[TPM_PT["PT_FIXED"] = 256] = "PT_FIXED";
    /**
     *  a 4-octet character string containing the TPM Family value (TPM_SPEC_FAMILY)
     */
    TPM_PT[TPM_PT["FAMILY_INDICATOR"] = 256] = "FAMILY_INDICATOR";
    /**
     *  the level of the specification
     *  NOTE 1	For this specification, the level is zero.
     *  NOTE 2	The level is on the title page of the specification.
     */
    TPM_PT[TPM_PT["LEVEL"] = 257] = "LEVEL";
    /**
     *  the specification Revision times 100
     *  EXAMPLE	Revision 01.01 would have a value of 101.
     *  NOTE	The Revision value is on the title page of the specification.
     */
    TPM_PT[TPM_PT["REVISION"] = 258] = "REVISION";
    /**
     *  the specification day of year using TCG calendar
     *  EXAMPLE	November 15, 2010, has a day of year value of 319 (0000013F16).
     *  NOTE The specification date is on the title page of the specification.
     */
    TPM_PT[TPM_PT["DAY_OF_YEAR"] = 259] = "DAY_OF_YEAR";
    /**
     *  the specification year using the CE
     *  EXAMPLE	The year 2010 has a value of 000007DA16.
     *  NOTE The specification date is on the title page of the specification.
     */
    TPM_PT[TPM_PT["YEAR"] = 260] = "YEAR";
    /**
     *  the vendor ID unique to each TPM manufacturer
     */
    TPM_PT[TPM_PT["MANUFACTURER"] = 261] = "MANUFACTURER";
    /**
     *  the first four characters of the vendor ID string
     *  NOTE	When the vendor string is fewer than 16 octets, the additional property values do not have to be present. A vendor string of 4 octets can be represented in one 32-bit value and no null terminating character is required.
     */
    TPM_PT[TPM_PT["VENDOR_STRING_1"] = 262] = "VENDOR_STRING_1";
    /**
     *  the second four characters of the vendor ID string
     */
    TPM_PT[TPM_PT["VENDOR_STRING_2"] = 263] = "VENDOR_STRING_2";
    /**
     *  the third four characters of the vendor ID string
     */
    TPM_PT[TPM_PT["VENDOR_STRING_3"] = 264] = "VENDOR_STRING_3";
    /**
     *  the fourth four characters of the vendor ID sting
     */
    TPM_PT[TPM_PT["VENDOR_STRING_4"] = 265] = "VENDOR_STRING_4";
    /**
     *  vendor-defined value indicating the TPM model
     */
    TPM_PT[TPM_PT["VENDOR_TPM_TYPE"] = 266] = "VENDOR_TPM_TYPE";
    /**
     *  the most-significant 32 bits of a TPM vendor-specific value indicating the version number of the firmware. See 10.12.2 and 10.12.8.
     */
    TPM_PT[TPM_PT["FIRMWARE_VERSION_1"] = 267] = "FIRMWARE_VERSION_1";
    /**
     *  the least-significant 32 bits of a TPM vendor-specific value indicating the version number of the firmware. See 10.12.2 and 10.12.8.
     */
    TPM_PT[TPM_PT["FIRMWARE_VERSION_2"] = 268] = "FIRMWARE_VERSION_2";
    /**
     *  the maximum size of a parameter (typically, a TPM2B_MAX_BUFFER)
     */
    TPM_PT[TPM_PT["INPUT_BUFFER"] = 269] = "INPUT_BUFFER";
    /**
     *  the minimum number of transient objects that can be held in TPM RAM
     *  NOTE	This minimum shall be no less than the minimum value required by the platform-specific specification to which the TPM is built.
     */
    TPM_PT[TPM_PT["HR_TRANSIENT_MIN"] = 270] = "HR_TRANSIENT_MIN";
    /**
     *  the minimum number of persistent objects that can be held in TPM NV memory
     *  NOTE	This minimum shall be no less than the minimum value required by the platform-specific specification to which the TPM is built.
     */
    TPM_PT[TPM_PT["HR_PERSISTENT_MIN"] = 271] = "HR_PERSISTENT_MIN";
    /**
     *  the minimum number of authorization sessions that can be held in TPM RAM
     *  NOTE	This minimum shall be no less than the minimum value required by the platform-specific specification to which the TPM is built.
     */
    TPM_PT[TPM_PT["HR_LOADED_MIN"] = 272] = "HR_LOADED_MIN";
    /**
     *  the number of authorization sessions that may be active at a time
     *  A session is active when it has a context associated with its handle. The context may either be in TPM RAM or be context saved.
     *  NOTE	This value shall be no less than the minimum value required by the platform-specific specification to which the TPM is built.
     */
    TPM_PT[TPM_PT["ACTIVE_SESSIONS_MAX"] = 273] = "ACTIVE_SESSIONS_MAX";
    /**
     *  the number of PCR implemented
     *  NOTE	This number is determined by the defined attributes, not the number of PCR that are populated.
     */
    TPM_PT[TPM_PT["PCR_COUNT"] = 274] = "PCR_COUNT";
    /**
     *  the minimum number of octets in a TPMS_PCR_SELECT.sizeOfSelect
     *  NOTE	This value is not determined by the number of PCR implemented but by the number of PCR required by the platform-specific specification with which the TPM is compliant or by the implementer if not adhering to a platform-specific specification.
     */
    TPM_PT[TPM_PT["PCR_SELECT_MIN"] = 275] = "PCR_SELECT_MIN";
    /**
     *  the maximum allowed difference (unsigned) between the contextID values of two saved session contexts
     *  This value shall be 2n-1, where n is at least 16.
     */
    TPM_PT[TPM_PT["CONTEXT_GAP_MAX"] = 276] = "CONTEXT_GAP_MAX";
    /**
     *  the maximum number of NV Indexes that are allowed to have the TPM_NT_COUNTER attribute
     *  NOTE	It is allowed for this value to be larger than the number of NV Indexes that can be defined. This would be indicative of a TPM implementation that did not use different implementation technology for different NV Index types.
     */
    TPM_PT[TPM_PT["NV_COUNTERS_MAX"] = 278] = "NV_COUNTERS_MAX";
    /**
     *  the maximum size of an NV Index data area
     */
    TPM_PT[TPM_PT["NV_INDEX_MAX"] = 279] = "NV_INDEX_MAX";
    /**
     *  a TPMA_MEMORY indicating the memory management method for the TPM
     */
    TPM_PT[TPM_PT["MEMORY"] = 280] = "MEMORY";
    /**
     *  interval, in milliseconds, between updates to the copy of TPMS_CLOCK_INFO.clock in NV
     */
    TPM_PT[TPM_PT["CLOCK_UPDATE"] = 281] = "CLOCK_UPDATE";
    /**
     *  the algorithm used for the integrity HMAC on saved contexts and for hashing the fuData of TPM2_FirmwareRead()
     */
    TPM_PT[TPM_PT["CONTEXT_HASH"] = 282] = "CONTEXT_HASH";
    /**
     *  TPM_ALG_ID, the algorithm used for encryption of saved contexts
     */
    TPM_PT[TPM_PT["CONTEXT_SYM"] = 283] = "CONTEXT_SYM";
    /**
     *  TPM_KEY_BITS, the size of the key used for encryption of saved contexts
     */
    TPM_PT[TPM_PT["CONTEXT_SYM_SIZE"] = 284] = "CONTEXT_SYM_SIZE";
    /**
     *  the modulus - 1 of the count for NV update of an orderly counter
     *  The returned value is MAX_ORDERLY_COUNT.
     *  This will have a value of 2N  1 where 1  N  32
     *  NOTE 1	An orderly counter is an NV Index with an TPM_NT of TPM_NV_COUNTER and TPMA_NV_ORDERLY SET.
     *  NOTE 2	When the low-order bits of a counter equal this value, an NV write occurs on the next increment.
     */
    TPM_PT[TPM_PT["ORDERLY_COUNT"] = 285] = "ORDERLY_COUNT";
    /**
     *  the maximum value for commandSize in a command
     */
    TPM_PT[TPM_PT["MAX_COMMAND_SIZE"] = 286] = "MAX_COMMAND_SIZE";
    /**
     *  the maximum value for responseSize in a response
     */
    TPM_PT[TPM_PT["MAX_RESPONSE_SIZE"] = 287] = "MAX_RESPONSE_SIZE";
    /**
     *  the maximum size of a digest that can be produced by the TPM
     */
    TPM_PT[TPM_PT["MAX_DIGEST"] = 288] = "MAX_DIGEST";
    /**
     *  the maximum size of an object context that will be returned by TPM2_ContextSave
     */
    TPM_PT[TPM_PT["MAX_OBJECT_CONTEXT"] = 289] = "MAX_OBJECT_CONTEXT";
    /**
     *  the maximum size of a session context that will be returned by TPM2_ContextSave
     */
    TPM_PT[TPM_PT["MAX_SESSION_CONTEXT"] = 290] = "MAX_SESSION_CONTEXT";
    /**
     *  platform-specific family (a TPM_PS value)(see Table 25)
     *  NOTE	The platform-specific values for the TPM_PT_PS parameters are in the relevant platform-specific specification. In the reference implementation, all of these values are 0.
     */
    TPM_PT[TPM_PT["PS_FAMILY_INDICATOR"] = 291] = "PS_FAMILY_INDICATOR";
    /**
     *  the level of the platform-specific specification
     */
    TPM_PT[TPM_PT["PS_LEVEL"] = 292] = "PS_LEVEL";
    /**
     *  the specification Revision times 100 for the platform-specific specification
     */
    TPM_PT[TPM_PT["PS_REVISION"] = 293] = "PS_REVISION";
    /**
     *  the platform-specific specification day of year using TCG calendar
     */
    TPM_PT[TPM_PT["PS_DAY_OF_YEAR"] = 294] = "PS_DAY_OF_YEAR";
    /**
     *  the platform-specific specification year using the CE
     */
    TPM_PT[TPM_PT["PS_YEAR"] = 295] = "PS_YEAR";
    /**
     *  the number of split signing operations supported by the TPM
     */
    TPM_PT[TPM_PT["SPLIT_MAX"] = 296] = "SPLIT_MAX";
    /**
     *  total number of commands implemented in the TPM
     */
    TPM_PT[TPM_PT["TOTAL_COMMANDS"] = 297] = "TOTAL_COMMANDS";
    /**
     *  number of commands from the TPM library that are implemented
     */
    TPM_PT[TPM_PT["LIBRARY_COMMANDS"] = 298] = "LIBRARY_COMMANDS";
    /**
     *  number of vendor commands that are implemented
     */
    TPM_PT[TPM_PT["VENDOR_COMMANDS"] = 299] = "VENDOR_COMMANDS";
    /**
     *  the maximum data size in one NV write, NV read, or NV certify command
     */
    TPM_PT[TPM_PT["NV_BUFFER_MAX"] = 300] = "NV_BUFFER_MAX";
    /**
     *  a TPMA_MODES value, indicating that the TPM is designed for these modes.
     */
    TPM_PT[TPM_PT["MODES"] = 301] = "MODES";
    /**
     *  the maximum size of a TPMS_CAPABILITY_DATA structure returned in TPM2_GetCapability().
     */
    TPM_PT[TPM_PT["MAX_CAP_BUFFER"] = 302] = "MAX_CAP_BUFFER";
    /**
     *  the group of variable properties returned as TPMS_TAGGED_PROPERTY
     *  The properties in this group change because of a Protected Capability other than a firmware update. The values are not necessarily persistent across all power transitions.
     */
    TPM_PT[TPM_PT["PT_VAR"] = 512] = "PT_VAR";
    /**
     *  TPMA_PERMANENT
     */
    TPM_PT[TPM_PT["PERMANENT"] = 512] = "PERMANENT";
    /**
     *  TPMA_STARTUP_CLEAR
     */
    TPM_PT[TPM_PT["STARTUP_CLEAR"] = 513] = "STARTUP_CLEAR";
    /**
     *  the number of NV Indexes currently defined
     */
    TPM_PT[TPM_PT["HR_NV_INDEX"] = 514] = "HR_NV_INDEX";
    /**
     *  the number of authorization sessions currently loaded into TPM RAM
     */
    TPM_PT[TPM_PT["HR_LOADED"] = 515] = "HR_LOADED";
    /**
     *  the number of additional authorization sessions, of any type, that could be loaded into TPM RAM
     *  This value is an estimate. If this value is at least 1, then at least one authorization session of any type may be loaded. Any command that changes the RAM memory allocation can make this estimate invalid.
     *  NOTE	A valid implementation may return 1 even if more than one authorization session would fit into RAM.
     */
    TPM_PT[TPM_PT["HR_LOADED_AVAIL"] = 516] = "HR_LOADED_AVAIL";
    /**
     *  the number of active authorization sessions currently being tracked by the TPM
     *  This is the sum of the loaded and saved sessions.
     */
    TPM_PT[TPM_PT["HR_ACTIVE"] = 517] = "HR_ACTIVE";
    /**
     *  the number of additional authorization sessions, of any type, that could be created
     *  This value is an estimate. If this value is at least 1, then at least one authorization session of any type may be created. Any command that changes the RAM memory allocation can make this estimate invalid.
     *  NOTE	A valid implementation may return 1 even if more than one authorization session could be created.
     */
    TPM_PT[TPM_PT["HR_ACTIVE_AVAIL"] = 518] = "HR_ACTIVE_AVAIL";
    /**
     *  estimate of the number of additional transient objects that could be loaded into TPM RAM
     *  This value is an estimate. If this value is at least 1, then at least one object of any type may be loaded. Any command that changes the memory allocation can make this estimate invalid.
     *  NOTE	A valid implementation may return 1 even if more than one transient object would fit into RAM.
     */
    TPM_PT[TPM_PT["HR_TRANSIENT_AVAIL"] = 519] = "HR_TRANSIENT_AVAIL";
    /**
     *  the number of persistent objects currently loaded into TPM NV memory
     */
    TPM_PT[TPM_PT["HR_PERSISTENT"] = 520] = "HR_PERSISTENT";
    /**
     *  the number of additional persistent objects that could be loaded into NV memory
     *  This value is an estimate. If this value is at least 1, then at least one object of any type may be made persistent. Any command that changes the NV memory allocation can make this estimate invalid.
     *  NOTE	A valid implementation may return 1 even if more than one persistent object would fit into NV memory.
     */
    TPM_PT[TPM_PT["HR_PERSISTENT_AVAIL"] = 521] = "HR_PERSISTENT_AVAIL";
    /**
     *  the number of defined NV Indexes that have NV the TPM_NT_COUNTER attribute
     */
    TPM_PT[TPM_PT["NV_COUNTERS"] = 522] = "NV_COUNTERS";
    /**
     *  the number of additional NV Indexes that can be defined with their TPM_NT of TPM_NV_COUNTER and the TPMA_NV_ORDERLY attribute SET
     *  This value is an estimate. If this value is at least 1, then at least one NV Index may be created with a TPM_NT of TPM_NV_COUNTER and the TPMA_NV_ORDERLY attributes. Any command that changes the NV memory allocation can make this estimate invalid.
     *  NOTE	A valid implementation may return 1 even if more than one NV counter could be defined.
     */
    TPM_PT[TPM_PT["NV_COUNTERS_AVAIL"] = 523] = "NV_COUNTERS_AVAIL";
    /**
     *  code that limits the algorithms that may be used with the TPM
     */
    TPM_PT[TPM_PT["ALGORITHM_SET"] = 524] = "ALGORITHM_SET";
    /**
     *  the number of loaded ECC curves
     */
    TPM_PT[TPM_PT["LOADED_CURVES"] = 525] = "LOADED_CURVES";
    /**
     *  the current value of the lockout counter (failedTries)
     */
    TPM_PT[TPM_PT["LOCKOUT_COUNTER"] = 526] = "LOCKOUT_COUNTER";
    /**
     *  the number of authorization failures before DA lockout is invoked
     */
    TPM_PT[TPM_PT["MAX_AUTH_FAIL"] = 527] = "MAX_AUTH_FAIL";
    /**
     *  the number of seconds before the value reported by TPM_PT_LOCKOUT_COUNTER is decremented
     */
    TPM_PT[TPM_PT["LOCKOUT_INTERVAL"] = 528] = "LOCKOUT_INTERVAL";
    /**
     *  the number of seconds after a lockoutAuth failure before use of lockoutAuth may be attempted again
     */
    TPM_PT[TPM_PT["LOCKOUT_RECOVERY"] = 529] = "LOCKOUT_RECOVERY";
    /**
     *  number of milliseconds before the TPM will accept another command that will modify NV
     *  This value is an approximation and may go up or down over time.
     */
    TPM_PT[TPM_PT["NV_WRITE_RECOVERY"] = 530] = "NV_WRITE_RECOVERY";
    /**
     *  the high-order 32 bits of the command audit counter
     */
    TPM_PT[TPM_PT["AUDIT_COUNTER_0"] = 531] = "AUDIT_COUNTER_0";
    /**
     *  the low-order 32 bits of the command audit counter
     */
    TPM_PT[TPM_PT["AUDIT_COUNTER_1"] = 532] = "AUDIT_COUNTER_1";
})(TPM_PT = exports.TPM_PT // UINT32
 || (exports.TPM_PT // UINT32
 = {}));
; // enum TPM_PT
/**
 *  The TPM_PT_PCR constants are used in TPM2_GetCapability() to indicate the property being selected or returned. The PCR properties can be read when capability == TPM_CAP_PCR_PROPERTIES. If there is no property that corresponds to the value of property, the next higher value is returned, if it exists.
 */
var TPM_PT_PCR;
(function (TPM_PT_PCR) {
    /**
     *  bottom of the range of TPM_PT_PCR properties
     */
    TPM_PT_PCR[TPM_PT_PCR["FIRST"] = 0] = "FIRST";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR is saved and restored by TPM_SU_STATE
     */
    TPM_PT_PCR[TPM_PT_PCR["SAVE"] = 0] = "SAVE";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be extended from locality 0
     *  This property is only present if a locality other than 0 is implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["EXTEND_L0"] = 1] = "EXTEND_L0";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be reset by TPM2_PCR_Reset() from locality 0
     */
    TPM_PT_PCR[TPM_PT_PCR["RESET_L0"] = 2] = "RESET_L0";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be extended from locality 1 This property is only present if locality 1 is implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["EXTEND_L1"] = 3] = "EXTEND_L1";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be reset by TPM2_PCR_Reset() from locality 1
     *  This property is only present if locality 1 is implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["RESET_L1"] = 4] = "RESET_L1";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be extended from locality 2 This property is only present if localities 1 and 2 are implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["EXTEND_L2"] = 5] = "EXTEND_L2";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be reset by TPM2_PCR_Reset() from locality 2
     *  This property is only present if localities 1 and 2 are implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["RESET_L2"] = 6] = "RESET_L2";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be extended from locality 3
     *  This property is only present if localities 1, 2, and 3 are implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["EXTEND_L3"] = 7] = "EXTEND_L3";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be reset by TPM2_PCR_Reset() from locality 3
     *  This property is only present if localities 1, 2, and 3 are implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["RESET_L3"] = 8] = "RESET_L3";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be extended from locality 4
     *  This property is only present if localities 1, 2, 3, and 4 are implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["EXTEND_L4"] = 9] = "EXTEND_L4";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR may be reset by TPM2_PCR_Reset() from locality 4
     *  This property is only present if localities 1, 2, 3, and 4 are implemented.
     */
    TPM_PT_PCR[TPM_PT_PCR["RESET_L4"] = 10] = "RESET_L4";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that modifications to this PCR (reset or Extend) will not increment the pcrUpdateCounter
     */
    TPM_PT_PCR[TPM_PT_PCR["NO_INCREMENT"] = 17] = "NO_INCREMENT";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR is reset by a D-RTM event
     *  These PCR are reset to -1 on TPM2_Startup() and reset to 0 on a _TPM_Hash_End event following a _TPM_Hash_Start event.
     */
    TPM_PT_PCR[TPM_PT_PCR["DRTM_RESET"] = 18] = "DRTM_RESET";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR is controlled by policy
     *  This property is only present if the TPM supports policy control of a PCR.
     */
    TPM_PT_PCR[TPM_PT_PCR["POLICY"] = 19] = "POLICY";
    /**
     *  a SET bit in the TPMS_PCR_SELECT indicates that the PCR is controlled by an authorization value
     *  This property is only present if the TPM supports authorization control of a PCR.
     */
    TPM_PT_PCR[TPM_PT_PCR["AUTH"] = 20] = "AUTH";
    /**
     *  top of the range of TPM_PT_PCR properties of the implementation
     *  If the TPM receives a request for a PCR property with a value larger than this, the TPM will return a zero length list and set the moreData parameter to NO.
     *  NOTE	This is an implementation-specific value. The value shown reflects the reference code implementation.
     */
    TPM_PT_PCR[TPM_PT_PCR["LAST"] = 20] = "LAST";
})(TPM_PT_PCR = exports.TPM_PT_PCR // UINT32
 || (exports.TPM_PT_PCR // UINT32
 = {}));
; // enum TPM_PT_PCR
/**
 *  The 32-bit handle space is divided into 256 regions of equal size with 224 values in each. Each of these ranges represents a handle type.
 */
var TPM_HT;
(function (TPM_HT) {
    /**
     *  PCR  consecutive numbers, starting at 0, that reference the PCR registers
     *  A platform-specific specification will set the minimum number of PCR and an implementation may have more.
     */
    TPM_HT[TPM_HT["PCR"] = 0] = "PCR";
    /**
     *  NV Index  assigned by the caller
     */
    TPM_HT[TPM_HT["NV_INDEX"] = 1] = "NV_INDEX";
    /**
     *  HMAC Authorization Session  assigned by the TPM when the session is created
     */
    TPM_HT[TPM_HT["HMAC_SESSION"] = 2] = "HMAC_SESSION";
    /**
     *  Loaded Authorization Session  used only in the context of TPM2_GetCapability
     *  This type references both loaded HMAC and loaded policy authorization sessions.
     */
    TPM_HT[TPM_HT["LOADED_SESSION"] = 2] = "LOADED_SESSION";
    /**
     *  Policy Authorization Session  assigned by the TPM when the session is created
     */
    TPM_HT[TPM_HT["POLICY_SESSION"] = 3] = "POLICY_SESSION";
    /**
     *  Saved Authorization Session  used only in the context of TPM2_GetCapability
     *  This type references saved authorization session contexts for which the TPM is maintaining tracking information.
     */
    TPM_HT[TPM_HT["SAVED_SESSION"] = 3] = "SAVED_SESSION";
    /**
     *  Permanent Values  assigned by this specification in Table 28
     */
    TPM_HT[TPM_HT["PERMANENT"] = 64] = "PERMANENT";
    /**
     *  Transient Objects  assigned by the TPM when an object is loaded into transient-object memory or when a persistent object is converted to a transient object
     */
    TPM_HT[TPM_HT["TRANSIENT"] = 128] = "TRANSIENT";
    /**
     *  Persistent Objects  assigned by the TPM when a loaded transient object is made persistent
     */
    TPM_HT[TPM_HT["PERSISTENT"] = 129] = "PERSISTENT";
})(TPM_HT = exports.TPM_HT // BYTE
 || (exports.TPM_HT // BYTE
 = {}));
; // enum TPM_HT
/**
 *  Table 28 lists the architecturally defined handles that cannot be changed. The handles include authorization handles, and special handles.
 */
var TPM_RH;
(function (TPM_RH) {
    TPM_RH[TPM_RH["FIRST"] = 1073741824] = "FIRST";
    /**
     *  not used1
     */
    TPM_RH[TPM_RH["SRK"] = 1073741824] = "SRK";
    /**
     *  handle references the Storage Primary Seed (SPS), the ownerAuth, and the ownerPolicy
     */
    TPM_RH[TPM_RH["OWNER"] = 1073741825] = "OWNER";
    /**
     *  not used1
     */
    TPM_RH[TPM_RH["REVOKE"] = 1073741826] = "REVOKE";
    /**
     *  not used1
     */
    TPM_RH[TPM_RH["TRANSPORT"] = 1073741827] = "TRANSPORT";
    /**
     *  not used1
     */
    TPM_RH[TPM_RH["OPERATOR"] = 1073741828] = "OPERATOR";
    /**
     *  not used1
     */
    TPM_RH[TPM_RH["ADMIN"] = 1073741829] = "ADMIN";
    /**
     *  not used1
     */
    TPM_RH[TPM_RH["EK"] = 1073741830] = "EK";
    /**
     *  a handle associated with the null hierarchy, an EmptyAuth authValue, and an Empty Policy authPolicy.
     */
    TPM_RH[TPM_RH["NULL"] = 1073741831] = "NULL";
    /**
     *  value reserved to the TPM to indicate a handle location that has not been initialized or assigned
     */
    TPM_RH[TPM_RH["UNASSIGNED"] = 1073741832] = "UNASSIGNED";
    /**
     *  authorization value used to indicate a password authorization session
     */
    TPM_RH[TPM_RH["RS_PW"] = 1073741833] = "RS_PW";
    /**
     *  references the authorization associated with the dictionary attack lockout reset
     */
    TPM_RH[TPM_RH["LOCKOUT"] = 1073741834] = "LOCKOUT";
    /**
     *  references the Endorsement Primary Seed (EPS), endorsementAuth, and endorsementPolicy
     */
    TPM_RH[TPM_RH["ENDORSEMENT"] = 1073741835] = "ENDORSEMENT";
    /**
     *  references the Platform Primary Seed (PPS), platformAuth, and platformPolicy
     */
    TPM_RH[TPM_RH["PLATFORM"] = 1073741836] = "PLATFORM";
    /**
     *  for phEnableNV
     */
    TPM_RH[TPM_RH["PLATFORM_NV"] = 1073741837] = "PLATFORM_NV";
    /**
     *  Start of a range of authorization values that are vendor-specific. A TPM may support any of the values in this range as are needed for vendor-specific purposes.
     *  Disabled if ehEnable is CLEAR.
     *  NOTE Any includes none.
     */
    TPM_RH[TPM_RH["AUTH_00"] = 1073741840] = "AUTH_00";
    /**
     *  End of the range of vendor-specific authorization values.
     */
    TPM_RH[TPM_RH["AUTH_FF"] = 1073742095] = "AUTH_FF";
    /**
     *  the top of the reserved handle area
     *  This is set to allow TPM2_GetCapability() to know where to stop. It may vary as implementations add to the permanent handle area.
     */
    TPM_RH[TPM_RH["LAST"] = 1073742095] = "LAST";
})(TPM_RH = exports.TPM_RH // UINT32
 || (exports.TPM_RH // UINT32
 = {}));
; // enum TPM_RH
/**
 *  This table lists the values of the TPM_NT field of a TPMA_NV. See Table 204 for usage.
 */
var TPM_NT;
(function (TPM_NT) {
    /**
     *  Ordinary  contains data that is opaque to the TPM that can only be modified using TPM2_NV_Write().
     */
    TPM_NT[TPM_NT["ORDINARY"] = 0] = "ORDINARY";
    /**
     *  Counter  contains an 8-octet value that is to be used as a counter and can only be modified with TPM2_NV_Increment()
     */
    TPM_NT[TPM_NT["COUNTER"] = 1] = "COUNTER";
    /**
     *  Bit Field  contains an 8-octet value to be used as a bit field and can only be modified with TPM2_NV_SetBits().
     */
    TPM_NT[TPM_NT["BITS"] = 2] = "BITS";
    /**
     *  Extend  contains a digest-sized value used like a PCR. The Index can only be modified using TPM2_NV_Extend(). The extend will use the nameAlg of the Index.
     */
    TPM_NT[TPM_NT["EXTEND"] = 4] = "EXTEND";
    /**
     *  PIN Fail - contains pinCount that increments on a PIN authorization failure and a pinLimit
     */
    TPM_NT[TPM_NT["PIN_FAIL"] = 8] = "PIN_FAIL";
    /**
     *  PIN Pass - contains pinCount that increments on a PIN authorization success and a pinLimit
     */
    TPM_NT[TPM_NT["PIN_PASS"] = 9] = "PIN_PASS";
})(TPM_NT = exports.TPM_NT // UINT32
 || (exports.TPM_NT // UINT32
 = {}));
; // enum TPM_NT
/**
 *  This table contains a collection of values used in various parts of the reference code. The values shown are illustrative.
 */
var Implementation;
(function (Implementation) {
    /**
     *  temporary define
     */
    //    FIELD_UPGRADE_IMPLEMENTED = Logic.NO,
    /**
     *  sets the size granularity for the buffers in a TPM2B structure
     *  TPMxB buffers will be assigned a space that is a multiple of this value. This does not set the size limits for IO. Those are set by the canonical form of the TPMxB
     */
    Implementation[Implementation["BUFFER_ALIGNMENT"] = 4] = "BUFFER_ALIGNMENT";
    /**
     *  the number of PCR in the TPM
     */
    Implementation[Implementation["IMPLEMENTATION_PCR"] = 24] = "IMPLEMENTATION_PCR";
    /**
     *  the number of PCR required by the relevant platform specification
     */
    Implementation[Implementation["PLATFORM_PCR"] = 24] = "PLATFORM_PCR";
    /**
     *  the D-RTM PCR
     *  NOTE This value is not defined when the TPM does not implement D-RTM
     */
    Implementation[Implementation["DRTM_PCR"] = 17] = "DRTM_PCR";
    /**
     *  the PCR that will receive the H-CRTM value at TPM2_Startup. This value should not be changed.
     */
    Implementation[Implementation["HCRTM_PCR"] = 0] = "HCRTM_PCR";
    /**
     *  the number of localities supported by the TPM
     *  This is expected to be either 5 for a PC, or 1 for just about everything else.
     */
    Implementation[Implementation["NUM_LOCALITIES"] = 5] = "NUM_LOCALITIES";
    /**
     *  the maximum number of handles in the handle area
     *  This should be produced by the Part 3 parser but is here for now.
     */
    Implementation[Implementation["MAX_HANDLE_NUM"] = 3] = "MAX_HANDLE_NUM";
    /**
     *  the number of simultaneously active sessions that are supported by the TPM implementation
     */
    Implementation[Implementation["MAX_ACTIVE_SESSIONS"] = 64] = "MAX_ACTIVE_SESSIONS";
    /**
     *  the number of sessions that the TPM may have in memory
     */
    Implementation[Implementation["MAX_LOADED_SESSIONS"] = 3] = "MAX_LOADED_SESSIONS";
    /**
     *  this is the current maximum value
     */
    Implementation[Implementation["MAX_SESSION_NUM"] = 3] = "MAX_SESSION_NUM";
    /**
     *  the number of simultaneously loaded objects that are supported by the TPM; this number does not include the objects that may be placed in NV memory by TPM2_EvictControl().
     */
    Implementation[Implementation["MAX_LOADED_OBJECTS"] = 3] = "MAX_LOADED_OBJECTS";
    /**
     *  the minimum number of evict objects supported by the TPM
     */
    Implementation[Implementation["MIN_EVICT_OBJECTS"] = 2] = "MIN_EVICT_OBJECTS";
    Implementation[Implementation["PCR_SELECT_MIN"] = 3.875] = "PCR_SELECT_MIN";
    Implementation[Implementation["PCR_SELECT_MAX"] = 3.875] = "PCR_SELECT_MAX";
    /**
     *  number of PCR groups that have individual policies
     */
    Implementation[Implementation["NUM_POLICY_PCR_GROUP"] = 1] = "NUM_POLICY_PCR_GROUP";
    /**
     *  number of PCR groups that have individual authorization values
     */
    Implementation[Implementation["NUM_AUTHVALUE_PCR_GROUP"] = 1] = "NUM_AUTHVALUE_PCR_GROUP";
    /**
     *  This may be larger than necessary
     */
    Implementation[Implementation["MAX_CONTEXT_SIZE"] = 2048] = "MAX_CONTEXT_SIZE";
    Implementation[Implementation["MAX_DIGEST_BUFFER"] = 1024] = "MAX_DIGEST_BUFFER";
    /**
     *  maximum data size allowed in an NV Index
     */
    Implementation[Implementation["MAX_NV_INDEX_SIZE"] = 2048] = "MAX_NV_INDEX_SIZE";
    /**
     *  maximum data size in one NV read or write command
     */
    Implementation[Implementation["MAX_NV_BUFFER_SIZE"] = 1024] = "MAX_NV_BUFFER_SIZE";
    /**
     *  maximum size of a capability buffer
     */
    Implementation[Implementation["MAX_CAP_BUFFER"] = 1024] = "MAX_CAP_BUFFER";
    /**
     *  size of NV memory in octets
     */
    Implementation[Implementation["NV_MEMORY_SIZE"] = 16384] = "NV_MEMORY_SIZE";
    Implementation[Implementation["NUM_STATIC_PCR"] = 16] = "NUM_STATIC_PCR";
    /**
     *  number of algorithms that can be in a list
     */
    Implementation[Implementation["MAX_ALG_LIST_SIZE"] = 64] = "MAX_ALG_LIST_SIZE";
    /**
     *  nominal value for the pre-scale value of Clock (the number of cycles of the TPM's oscillator for each increment of Clock)
     */
    Implementation[Implementation["TIMER_PRESCALE"] = 100000] = "TIMER_PRESCALE";
    /**
     *  size of the Primary Seed in octets
     */
    Implementation[Implementation["PRIMARY_SEED_SIZE"] = 32] = "PRIMARY_SEED_SIZE";
    /**
     *  context encryption algorithm
     */
    Implementation[Implementation["CONTEXT_ENCRYPT_ALG"] = 6] = "CONTEXT_ENCRYPT_ALG";
    /**
     *  context encryption key size in bits
     */
    //    CONTEXT_ENCRYPT_KEY_BITS = ImplementationConstants.MAX_SYM_KEY_BITS,
    //    CONTEXT_ENCRYPT_KEY_BYTES = ((Implementation.CONTEXT_ENCRYPT_KEY_BITS+7)/8),
    /**
     *  context integrity hash algorithm
     */
    Implementation[Implementation["CONTEXT_INTEGRITY_HASH_ALG"] = 11] = "CONTEXT_INTEGRITY_HASH_ALG";
    /**
     *  number of byes in the context integrity digest
     */
    //    CONTEXT_INTEGRITY_HASH_SIZE = UINT32(SHA256::DIGEST_SIZE),
    /**
     *  size of proof value in octets
     *  This size of the proof should be consistent with the digest size used for context integrity.
     */
    //    PROOF_SIZE = Implementation.CONTEXT_INTEGRITY_HASH_SIZE,
    /**
     *  the update interval expressed as a power of 2 seconds
     *  A value of 12 is 4,096 seconds (~68 minutes).
     */
    Implementation[Implementation["NV_CLOCK_UPDATE_INTERVAL"] = 12] = "NV_CLOCK_UPDATE_INTERVAL";
    /**
     *  number of PCR groups that allow policy/auth
     */
    Implementation[Implementation["NUM_POLICY_PCR"] = 1] = "NUM_POLICY_PCR";
    /**
     *  maximum size of a command
     */
    Implementation[Implementation["MAX_COMMAND_SIZE"] = 4096] = "MAX_COMMAND_SIZE";
    /**
     *  maximum size of a response
     */
    Implementation[Implementation["MAX_RESPONSE_SIZE"] = 4096] = "MAX_RESPONSE_SIZE";
    /**
     *  number between 1 and 32 inclusive
     */
    Implementation[Implementation["ORDERLY_BITS"] = 8] = "ORDERLY_BITS";
    /**
     *  maximum count of orderly counter before NV is updated
     *  This must be of the form 2N  1 where 1  N  32.
     */
    Implementation[Implementation["MAX_ORDERLY_COUNT"] = 255] = "MAX_ORDERLY_COUNT";
    /**
     *  used by TPM2_GetCapability() processing to bound the algorithm search
     */
    Implementation[Implementation["ALG_ID_FIRST"] = 1] = "ALG_ID_FIRST";
    /**
     *  used by TPM2_GetCapability() processing to bound the algorithm search
     */
    Implementation[Implementation["ALG_ID_LAST"] = 68] = "ALG_ID_LAST";
    /**
     *  the maximum number of octets that may be in a sealed blob; 128 is the minimum allowed value
     */
    Implementation[Implementation["MAX_SYM_DATA"] = 128] = "MAX_SYM_DATA";
    Implementation[Implementation["MAX_RNG_ENTROPY_SIZE"] = 64] = "MAX_RNG_ENTROPY_SIZE";
    Implementation[Implementation["RAM_INDEX_SPACE"] = 512] = "RAM_INDEX_SPACE";
    /**
     *  216 + 1
     */
    Implementation[Implementation["RSA_DEFAULT_PUBLIC_EXPONENT"] = 65537] = "RSA_DEFAULT_PUBLIC_EXPONENT";
    /**
     *  indicates if the TPM_PT_PCR_NO_INCREMENT group is implemented
     */
    //    ENABLE_PCR_NO_INCREMENT = Logic.YES,
    //    CRT_FORMAT_RSA = Logic.YES,
    Implementation[Implementation["VENDOR_COMMAND_COUNT"] = 0] = "VENDOR_COMMAND_COUNT";
    /**
     *  MAX_RSA_KEY_BYTES is auto generated from the RSA key size selection in Table 4. If RSA is not implemented, this may need to be manually removed.
     */
    //    PRIVATE_VENDOR_SPECIFIC_BYTES = ((ImplementationConstants.MAX_RSA_KEY_BYTES/2) * (3 + Implementation.CRT_FORMAT_RSA * 2)),
    /**
     *  Maximum size of the vendor-specific buffer
     */
    Implementation[Implementation["MAX_VENDOR_BUFFER_SIZE"] = 1024] = "MAX_VENDOR_BUFFER_SIZE";
})(Implementation = exports.Implementation // UINT32
 || (exports.Implementation // UINT32
 = {}));
; // enum Implementation
/**
 *  The definitions in Table 29 are used to define many of the interface data types.
 */
var TPM_HC;
(function (TPM_HC) {
    /**
     *  to mask off the HR
     */
    TPM_HC[TPM_HC["HR_HANDLE_MASK"] = 16777215] = "HR_HANDLE_MASK";
    /**
     *  to mask off the variable part
     */
    TPM_HC[TPM_HC["HR_RANGE_MASK"] = 4278190080] = "HR_RANGE_MASK";
    TPM_HC[TPM_HC["HR_SHIFT"] = 24] = "HR_SHIFT";
    TPM_HC[TPM_HC["HR_PCR"] = 0] = "HR_PCR";
    TPM_HC[TPM_HC["HR_HMAC_SESSION"] = 33554432] = "HR_HMAC_SESSION";
    TPM_HC[TPM_HC["HR_POLICY_SESSION"] = 50331648] = "HR_POLICY_SESSION";
    TPM_HC[TPM_HC["HR_TRANSIENT"] = -2147483648] = "HR_TRANSIENT";
    TPM_HC[TPM_HC["HR_PERSISTENT"] = -2130706432] = "HR_PERSISTENT";
    TPM_HC[TPM_HC["HR_NV_INDEX"] = 16777216] = "HR_NV_INDEX";
    TPM_HC[TPM_HC["HR_PERMANENT"] = 1073741824] = "HR_PERMANENT";
    /**
     *  first PCR
     */
    TPM_HC[TPM_HC["PCR_FIRST"] = 0] = "PCR_FIRST";
    /**
     *  last PCR
     */
    TPM_HC[TPM_HC["PCR_LAST"] = 23] = "PCR_LAST";
    /**
     *  first HMAC session
     */
    TPM_HC[TPM_HC["HMAC_SESSION_FIRST"] = 33554432] = "HMAC_SESSION_FIRST";
    /**
     *  last HMAC session
     */
    TPM_HC[TPM_HC["HMAC_SESSION_LAST"] = 33554495] = "HMAC_SESSION_LAST";
    /**
     *  used in GetCapability
     */
    TPM_HC[TPM_HC["LOADED_SESSION_FIRST"] = 33554432] = "LOADED_SESSION_FIRST";
    /**
     *  used in GetCapability
     */
    TPM_HC[TPM_HC["LOADED_SESSION_LAST"] = 33554495] = "LOADED_SESSION_LAST";
    /**
     *  first policy session
     */
    TPM_HC[TPM_HC["POLICY_SESSION_FIRST"] = 50331648] = "POLICY_SESSION_FIRST";
    /**
     *  last policy session
     */
    TPM_HC[TPM_HC["POLICY_SESSION_LAST"] = 50331711] = "POLICY_SESSION_LAST";
    /**
     *  first transient object
     */
    TPM_HC[TPM_HC["TRANSIENT_FIRST"] = -2147483648] = "TRANSIENT_FIRST";
    /**
     *  used in GetCapability
     */
    TPM_HC[TPM_HC["ACTIVE_SESSION_FIRST"] = 50331648] = "ACTIVE_SESSION_FIRST";
    /**
     *  used in GetCapability
     */
    TPM_HC[TPM_HC["ACTIVE_SESSION_LAST"] = 50331711] = "ACTIVE_SESSION_LAST";
    /**
     *  last transient object
     */
    TPM_HC[TPM_HC["TRANSIENT_LAST"] = -2147483646] = "TRANSIENT_LAST";
    /**
     *  first persistent object
     */
    TPM_HC[TPM_HC["PERSISTENT_FIRST"] = -2130706432] = "PERSISTENT_FIRST";
    /**
     *  last persistent object
     */
    TPM_HC[TPM_HC["PERSISTENT_LAST"] = -2113929217] = "PERSISTENT_LAST";
    /**
     *  first platform persistent object
     */
    TPM_HC[TPM_HC["PLATFORM_PERSISTENT"] = -2122317824] = "PLATFORM_PERSISTENT";
    /**
     *  first allowed NV Index
     */
    TPM_HC[TPM_HC["NV_INDEX_FIRST"] = 16777216] = "NV_INDEX_FIRST";
    /**
     *  last allowed NV Index
     */
    TPM_HC[TPM_HC["NV_INDEX_LAST"] = 33554431] = "NV_INDEX_LAST";
    TPM_HC[TPM_HC["PERMANENT_FIRST"] = 1073741824] = "PERMANENT_FIRST";
    TPM_HC[TPM_HC["PERMANENT_LAST"] = 1073742095] = "PERMANENT_LAST";
})(TPM_HC = exports.TPM_HC // UINT32
 || (exports.TPM_HC // UINT32
 = {}));
; // enum TPM_HC
/**
 *  This structure defines the attributes of an algorithm.
 */
var TPMA_ALGORITHM;
(function (TPMA_ALGORITHM) {
    /**
     *  SET (1): an asymmetric algorithm with public and private portions
     *  CLEAR (0): not an asymmetric algorithm
     */
    TPMA_ALGORITHM[TPMA_ALGORITHM["asymmetric"] = 1] = "asymmetric";
    /**
     *  SET (1): a symmetric block cipher
     *  CLEAR (0): not a symmetric block cipher
     */
    TPMA_ALGORITHM[TPMA_ALGORITHM["symmetric"] = 2] = "symmetric";
    /**
     *  SET (1): a hash algorithm
     *  CLEAR (0): not a hash algorithm
     */
    TPMA_ALGORITHM[TPMA_ALGORITHM["hash"] = 4] = "hash";
    /**
     *  SET (1): an algorithm that may be used as an object type
     *  CLEAR (0): an algorithm that is not used as an object type
     */
    TPMA_ALGORITHM[TPMA_ALGORITHM["object"] = 8] = "object";
    /**
     *  SET (1): a signing algorithm. The setting of asymmetric, symmetric, and hash will indicate the type of signing algorithm.
     *  CLEAR (0): not a signing algorithm
     */
    TPMA_ALGORITHM[TPMA_ALGORITHM["signing"] = 256] = "signing";
    /**
     *  SET (1): an encryption/decryption algorithm. The setting of asymmetric, symmetric, and hash will indicate the type of encryption/decryption algorithm.
     *  CLEAR (0): not an encryption/decryption algorithm
     */
    TPMA_ALGORITHM[TPMA_ALGORITHM["encrypting"] = 512] = "encrypting";
    /**
     *  SET (1): a method such as a key derivative function (KDF)
     *  CLEAR (0): not a method
     */
    TPMA_ALGORITHM[TPMA_ALGORITHM["method"] = 1024] = "method";
})(TPMA_ALGORITHM = exports.TPMA_ALGORITHM // UINT32
 || (exports.TPMA_ALGORITHM // UINT32
 = {}));
; // enum TPMA_ALGORITHM
/**
 *  This attribute structure indicates an objects use, its authorization types, and its relationship to other objects.
 */
var TPMA_OBJECT;
(function (TPMA_OBJECT) {
    /**
     *  SET (1): The hierarchy of the object, as indicated by its Qualified Name, may not change.
     *  CLEAR (0): The hierarchy of the object may change as a result of this object or an ancestor key being duplicated for use in another hierarchy.
     */
    TPMA_OBJECT[TPMA_OBJECT["fixedTPM"] = 2] = "fixedTPM";
    /**
     *  SET (1): Previously saved contexts of this object may not be loaded after Startup(CLEAR).
     *  CLEAR (0): Saved contexts of this object may be used after a Shutdown(STATE) and subsequent Startup().
     */
    TPMA_OBJECT[TPMA_OBJECT["stClear"] = 4] = "stClear";
    /**
     *  SET (1): The parent of the object may not change.
     *  CLEAR (0): The parent of the object may change as the result of a TPM2_Duplicate() of the object.
     */
    TPMA_OBJECT[TPMA_OBJECT["fixedParent"] = 16] = "fixedParent";
    /**
     *  SET (1): Indicates that, when the object was created with TPM2_Create() or TPM2_CreatePrimary(), the TPM generated all of the sensitive data other than the authValue.
     *  CLEAR (0): A portion of the sensitive data, other than the authValue, was provided by the caller.
     */
    TPMA_OBJECT[TPMA_OBJECT["sensitiveDataOrigin"] = 32] = "sensitiveDataOrigin";
    /**
     *  SET (1): Approval of USER role actions with this object may be with an HMAC session or with a password using the authValue of the object or a policy session.
     *  CLEAR (0): Approval of USER role actions with this object may only be done with a policy session.
     */
    TPMA_OBJECT[TPMA_OBJECT["userWithAuth"] = 64] = "userWithAuth";
    /**
     *  SET (1): Approval of ADMIN role actions with this object may only be done with a policy session.
     *  CLEAR (0): Approval of ADMIN role actions with this object may be with an HMAC session or with a password using the authValue of the object or a policy session.
     */
    TPMA_OBJECT[TPMA_OBJECT["adminWithPolicy"] = 128] = "adminWithPolicy";
    /**
     *  SET (1): The object is not subject to dictionary attack protections.
     *  CLEAR (0): The object is subject to dictionary attack protections.
     */
    TPMA_OBJECT[TPMA_OBJECT["noDA"] = 1024] = "noDA";
    /**
     *  SET (1): If the object is duplicated, then symmetricAlg shall not be TPM_ALG_NULL and newParentHandle shall not be TPM_RH_NULL.
     *  CLEAR (0): The object may be duplicated without an inner wrapper on the private portion of the object and the new parent may be TPM_RH_NULL.
     */
    TPMA_OBJECT[TPMA_OBJECT["encryptedDuplication"] = 2048] = "encryptedDuplication";
    /**
     *  SET (1): Key usage is restricted to manipulate structures of known format; the parent of this key shall have restricted SET.
     *  CLEAR (0): Key usage is not restricted to use on special formats.
     */
    TPMA_OBJECT[TPMA_OBJECT["restricted"] = 65536] = "restricted";
    /**
     *  SET (1): The private portion of the key may be used to decrypt.
     *  CLEAR (0): The private portion of the key may not be used to decrypt.
     */
    TPMA_OBJECT[TPMA_OBJECT["decrypt"] = 131072] = "decrypt";
    /**
     *  SET (1): For a symmetric cipher object, the private portion of the key may be used to encrypt. For other objects, the private portion of the key may be used to sign.
     *  CLEAR (0): The private portion of the key may not be used to sign or encrypt.
     */
    TPMA_OBJECT[TPMA_OBJECT["sign"] = 262144] = "sign";
    /**
     *  Alias to the Sign value.
     */
    TPMA_OBJECT[TPMA_OBJECT["encrypt"] = 262144] = "encrypt";
})(TPMA_OBJECT = exports.TPMA_OBJECT // UINT32
 || (exports.TPMA_OBJECT // UINT32
 = {}));
; // enum TPMA_OBJECT
/**
 *  This octet in each session is used to identify the session type, indicate its relationship to any handles in the command, and indicate its use in parameter encryption.
 */
var TPMA_SESSION;
(function (TPMA_SESSION) {
    /**
     *  SET (1): In a command, this setting indicates that the session is to remain active after successful completion of the command. In a response, it indicates that the session is still active. If SET in the command, this attribute shall be SET in the response.
     *  CLEAR (0): In a command, this setting indicates that the TPM should close the session and flush any related context when the command completes successfully. In a response, it indicates that the session is closed and the context is no longer active.
     *  This attribute has no meaning for a password authorization and the TPM will allow any setting of the attribute in the command and SET the attribute in the response.
     *  This attribute will only be CLEAR in one response for a logical session. If the attribute is CLEAR, the context associated with the session is no longer in use and the space is available. A session created after another session is ended may have the same handle but logically is not the same session.
     *  This attribute has no effect if the command does not complete successfully.
     */
    TPMA_SESSION[TPMA_SESSION["continueSession"] = 1] = "continueSession";
    /**
     *  SET (1): In a command, this setting indicates that the command should only be executed if the session is exclusive at the start of the command. In a response, it indicates that the session is exclusive. This setting is only allowed if the audit attribute is SET (TPM_RC_ATTRIBUTES).
     *  CLEAR (0): In a command, indicates that the session need not be exclusive at the start of the command. In a response, indicates that the session is not exclusive.
     *  In this revision, if audit is CLEAR, auditExclusive must be CLEAR in the command and will be CLEAR in the response. In a future, revision, this bit may have a different meaning if audit is CLEAR.
     *  See "Exclusive Audit Session" clause in TPM 2.0 Part 1.
     */
    TPMA_SESSION[TPMA_SESSION["auditExclusive"] = 2] = "auditExclusive";
    /**
     *  SET (1): In a command, this setting indicates that the audit digest of the session should be initialized and the exclusive status of the session SET. This setting is only allowed if the audit attribute is SET (TPM_RC_ATTRIBUTES).
     *  CLEAR (0): In a command, indicates that the audit digest should not be initialized.
     *  This bit is always CLEAR in a response.
     *  In this revision, if audit is CLEAR, auditReset must be clear in the command and will be CLEAR in the response. In a future, revision, this bit may have a different meaning if audit is CLEAR.
     */
    TPMA_SESSION[TPMA_SESSION["auditReset"] = 4] = "auditReset";
    /**
     *  SET (1): In a command, this setting indicates that the first parameter in the command is symmetrically encrypted using the parameter encryption scheme described in TPM 2.0 Part 1. The TPM will decrypt the parameter after performing any HMAC computations and before unmarshaling the parameter. In a response, the attribute is copied from the request but has no effect on the response.
     *  CLEAR (0): Session not used for encryption.
     *  For a password authorization, this attribute will be CLEAR in both the command and response.
     *  This attribute may only be SET in one session per command.
     *  This attribute may be SET in a session that is not associated with a command handle. Such a session is provided for purposes of encrypting a parameter and not for authorization.
     *  This attribute may be SET in combination with any other session attributes.
     *  This attribute may only be SET if the first parameter of the command is a sized buffer (TPM2B_).
     */
    TPMA_SESSION[TPMA_SESSION["decrypt"] = 32] = "decrypt";
    /**
     *  SET (1): In a command, this setting indicates that the TPM should use this session to encrypt the first parameter in the response. In a response, it indicates that the attribute was set in the command and that the TPM used the session to encrypt the first parameter in the response using the parameter encryption scheme described in TPM 2.0 Part 1.
     *  CLEAR (0): Session not used for encryption.
     *  For a password authorization, this attribute will be CLEAR in both the command and response.
     *  This attribute may only be SET in one session per command.
     *  This attribute may be SET in a session that is not associated with a command handle. Such a session is provided for purposes of encrypting a parameter and not for authorization.
     *  This attribute may only be SET if the first parameter of a response is a sized buffer (TPM2B_).
     */
    TPMA_SESSION[TPMA_SESSION["encrypt"] = 64] = "encrypt";
    /**
     *  SET (1): In a command or response, this setting indicates that the session is for audit and that auditExclusive and auditReset have meaning. This session may also be used for authorization, encryption, or decryption. The encrypted and encrypt fields may be SET or CLEAR.
     *  CLEAR (0): Session is not used for audit.
     *  This attribute may only be SET in one session per command or response. If SET in the command, then this attribute will be SET in the response.
     */
    TPMA_SESSION[TPMA_SESSION["audit"] = 128] = "audit";
})(TPMA_SESSION = exports.TPMA_SESSION // BYTE
 || (exports.TPMA_SESSION // BYTE
 = {}));
; // enum TPMA_SESSION
/**
 *  In a TPMS_CREATION_DATA structure, this structure is used to indicate the locality of the command that created the object. No more than one of the locality attributes shall be set in the creation data.
 */
var TPMA_LOCALITY;
(function (TPMA_LOCALITY) {
    TPMA_LOCALITY[TPMA_LOCALITY["LOC_ZERO"] = 1] = "LOC_ZERO";
    TPMA_LOCALITY[TPMA_LOCALITY["LOC_ONE"] = 2] = "LOC_ONE";
    TPMA_LOCALITY[TPMA_LOCALITY["LOC_TWO"] = 4] = "LOC_TWO";
    TPMA_LOCALITY[TPMA_LOCALITY["LOC_THREE"] = 8] = "LOC_THREE";
    TPMA_LOCALITY[TPMA_LOCALITY["LOC_FOUR"] = 16] = "LOC_FOUR";
    /**
     *  If any of these bits is set, an extended locality is indicated
     */
    TPMA_LOCALITY[TPMA_LOCALITY["Extended_BIT_MASK"] = 224] = "Extended_BIT_MASK";
    TPMA_LOCALITY[TPMA_LOCALITY["Extended_BIT_OFFSET"] = 5] = "Extended_BIT_OFFSET";
    TPMA_LOCALITY[TPMA_LOCALITY["Extended_BIT_LENGTH"] = 3] = "Extended_BIT_LENGTH";
    TPMA_LOCALITY[TPMA_LOCALITY["Extended_BIT_0"] = 32] = "Extended_BIT_0";
    TPMA_LOCALITY[TPMA_LOCALITY["Extended_BIT_1"] = 64] = "Extended_BIT_1";
    TPMA_LOCALITY[TPMA_LOCALITY["Extended_BIT_2"] = 128] = "Extended_BIT_2";
})(TPMA_LOCALITY = exports.TPMA_LOCALITY // BYTE
 || (exports.TPMA_LOCALITY // BYTE
 = {}));
; // enum TPMA_LOCALITY
/**
 *  The attributes in this structure are persistent and are not changed as a result of _TPM_Init or any TPM2_Startup(). Some of the attributes in this structure may change as the result of specific Protected Capabilities. This structure may be read using TPM2_GetCapability(capability = TPM_CAP_TPM_PROPERTIES, property = TPM_PT_PERMANENT).
 */
var TPMA_PERMANENT;
(function (TPMA_PERMANENT) {
    /**
     *  SET (1): TPM2_HierarchyChangeAuth() with ownerAuth has been executed since the last TPM2_Clear().
     *  CLEAR (0): ownerAuth has not been changed since TPM2_Clear().
     */
    TPMA_PERMANENT[TPMA_PERMANENT["ownerAuthSet"] = 1] = "ownerAuthSet";
    /**
     *  SET (1): TPM2_HierarchyChangeAuth() with endorsementAuth has been executed since the last TPM2_Clear().
     *  CLEAR (0): endorsementAuth has not been changed since TPM2_Clear().
     */
    TPMA_PERMANENT[TPMA_PERMANENT["endorsementAuthSet"] = 2] = "endorsementAuthSet";
    /**
     *  SET (1): TPM2_HierarchyChangeAuth() with lockoutAuth has been executed since the last TPM2_Clear().
     *  CLEAR (0): lockoutAuth has not been changed since TPM2_Clear().
     */
    TPMA_PERMANENT[TPMA_PERMANENT["lockoutAuthSet"] = 4] = "lockoutAuthSet";
    /**
     *  SET (1): TPM2_Clear() is disabled.
     *  CLEAR (0): TPM2_Clear() is enabled.
     *  NOTE	See TPM2_ClearControl in TPM 2.0 Part 3 for details on changing this attribute.
     */
    TPMA_PERMANENT[TPMA_PERMANENT["disableClear"] = 256] = "disableClear";
    /**
     *  SET (1): The TPM is in lockout, when failedTries is equal to maxTries.
     */
    TPMA_PERMANENT[TPMA_PERMANENT["inLockout"] = 512] = "inLockout";
    /**
     *  SET (1): The EPS was created by the TPM.
     *  CLEAR (0): The EPS was created outside of the TPM using a manufacturer-specific process.
     */
    TPMA_PERMANENT[TPMA_PERMANENT["tpmGeneratedEPS"] = 1024] = "tpmGeneratedEPS";
})(TPMA_PERMANENT = exports.TPMA_PERMANENT // UINT32
 || (exports.TPMA_PERMANENT // UINT32
 = {}));
; // enum TPMA_PERMANENT
/**
 *  This structure may be read using TPM2_GetCapability(capability = TPM_CAP_TPM_PROPERTIES, property = TPM_PT_STARTUP_CLEAR).
 */
var TPMA_STARTUP_CLEAR;
(function (TPMA_STARTUP_CLEAR) {
    /**
     *  SET (1): The platform hierarchy is enabled and platformAuth or platformPolicy may be used for authorization.
     *  CLEAR (0): platformAuth and platformPolicy may not be used for authorizations, and objects in the platform hierarchy, including persistent objects, cannot be used.
     *  NOTE	See TPM2_HierarchyControl in TPM 2.0 Part 3 for details on changing this attribute.
     */
    TPMA_STARTUP_CLEAR[TPMA_STARTUP_CLEAR["phEnable"] = 1] = "phEnable";
    /**
     *  SET (1): The Storage hierarchy is enabled and ownerAuth or ownerPolicy may be used for authorization. NV indices defined using owner authorization are accessible.
     *  CLEAR (0): ownerAuth and ownerPolicy may not be used for authorizations, and objects in the Storage hierarchy, persistent objects, and NV indices defined using owner authorization cannot be used.
     *  NOTE	See TPM2_HierarchyControl in TPM 2.0 Part 3 for details on changing this attribute.
     */
    TPMA_STARTUP_CLEAR[TPMA_STARTUP_CLEAR["shEnable"] = 2] = "shEnable";
    /**
     *  SET (1): The EPS hierarchy is enabled and Endorsement Authorization may be used to authorize commands.
     *  CLEAR (0): Endorsement Authorization may not be used for authorizations, and objects in the endorsement hierarchy, including persistent objects, cannot be used.
     *  NOTE	See TPM2_HierarchyControl in TPM 2.0 Part 3 for details on changing this attribute.
     */
    TPMA_STARTUP_CLEAR[TPMA_STARTUP_CLEAR["ehEnable"] = 4] = "ehEnable";
    /**
     *  SET (1): NV indices that have TPMA_PLATFORM_CREATE SET may be read or written. The platform can create define and undefine indices.
     *  CLEAR (0): NV indices that have TPMA_PLATFORM_CREATE SET may not be read or written (TPM_RC_HANDLE). The platform cannot define (TPM_RC_HIERARCHY) or undefined (TPM_RC_HANDLE) indices.
     *  NOTE	See TPM2_HierarchyControl in TPM 2.0 Part 3 for details on changing this attribute.
     *  NOTE read refers to these commands: TPM2_NV_Read, TPM2_NV_ReadPublic, TPM_NV_Certify, TPM2_PolicyNV
     *  write refers to these commands: TPM2_NV_Write, TPM2_NV_Increment, TPM2_NV_Extend, TPM2_NV_SetBits
     *  NOTE The TPM must query the index TPMA_PLATFORM_CREATE attribute to determine whether phEnableNV is applicable. Since the TPM will return TPM_RC_HANDLE if the index does not exist, it also returns this error code if the index is disabled. Otherwise, the TPM would leak the existence of an index even when disabled.
     */
    TPMA_STARTUP_CLEAR[TPMA_STARTUP_CLEAR["phEnableNV"] = 8] = "phEnableNV";
    /**
     *  SET (1): The TPM received a TPM2_Shutdown() and a matching TPM2_Startup().
     *  CLEAR (0): TPM2_Startup(TPM_SU_CLEAR) was not preceded by a TPM2_Shutdown() of any type.
     *  NOTE A shutdown is orderly if the TPM receives a TPM2_Shutdown() of any type followed by a TPM2_Startup() of any type. However, the TPM will return an error if TPM2_Startup(TPM_SU_STATE) was not preceded by TPM2_Shutdown(TPM_SU_STATE).
     */
    TPMA_STARTUP_CLEAR[TPMA_STARTUP_CLEAR["orderly"] = 2147483648] = "orderly";
})(TPMA_STARTUP_CLEAR = exports.TPMA_STARTUP_CLEAR // UINT32
 || (exports.TPMA_STARTUP_CLEAR // UINT32
 = {}));
; // enum TPMA_STARTUP_CLEAR
/**
 *  This structure defines the attributes of a command from a context management perspective. The fields of the structure indicate to the TPM Resource Manager (TRM) the number of resources required by a command and how the command affects the TPMs resources.
 */
var TPMA_CC;
(function (TPMA_CC) {
    /**
     *  indicates the command being selected
     */
    TPMA_CC[TPMA_CC["commandIndex_BIT_MASK"] = 65535] = "commandIndex_BIT_MASK";
    TPMA_CC[TPMA_CC["commandIndex_BIT_OFFSET"] = 0] = "commandIndex_BIT_OFFSET";
    TPMA_CC[TPMA_CC["commandIndex_BIT_LENGTH"] = 16] = "commandIndex_BIT_LENGTH";
    TPMA_CC[TPMA_CC["commandIndex_BIT_0"] = 1] = "commandIndex_BIT_0";
    TPMA_CC[TPMA_CC["commandIndex_BIT_1"] = 2] = "commandIndex_BIT_1";
    TPMA_CC[TPMA_CC["commandIndex_BIT_2"] = 4] = "commandIndex_BIT_2";
    TPMA_CC[TPMA_CC["commandIndex_BIT_3"] = 8] = "commandIndex_BIT_3";
    TPMA_CC[TPMA_CC["commandIndex_BIT_4"] = 16] = "commandIndex_BIT_4";
    TPMA_CC[TPMA_CC["commandIndex_BIT_5"] = 32] = "commandIndex_BIT_5";
    TPMA_CC[TPMA_CC["commandIndex_BIT_6"] = 64] = "commandIndex_BIT_6";
    TPMA_CC[TPMA_CC["commandIndex_BIT_7"] = 128] = "commandIndex_BIT_7";
    TPMA_CC[TPMA_CC["commandIndex_BIT_8"] = 256] = "commandIndex_BIT_8";
    TPMA_CC[TPMA_CC["commandIndex_BIT_9"] = 512] = "commandIndex_BIT_9";
    TPMA_CC[TPMA_CC["commandIndex_BIT_10"] = 1024] = "commandIndex_BIT_10";
    TPMA_CC[TPMA_CC["commandIndex_BIT_11"] = 2048] = "commandIndex_BIT_11";
    TPMA_CC[TPMA_CC["commandIndex_BIT_12"] = 4096] = "commandIndex_BIT_12";
    TPMA_CC[TPMA_CC["commandIndex_BIT_13"] = 8192] = "commandIndex_BIT_13";
    TPMA_CC[TPMA_CC["commandIndex_BIT_14"] = 16384] = "commandIndex_BIT_14";
    TPMA_CC[TPMA_CC["commandIndex_BIT_15"] = 32768] = "commandIndex_BIT_15";
    /**
     *  SET (1): indicates that the command may write to NV
     *  CLEAR (0): indicates that the command does not write to NV
     */
    TPMA_CC[TPMA_CC["nv"] = 4194304] = "nv";
    /**
     *  SET (1): This command could flush any number of loaded contexts.
     *  CLEAR (0): no additional changes other than indicated by the flushed attribute
     */
    TPMA_CC[TPMA_CC["extensive"] = 8388608] = "extensive";
    /**
     *  SET (1): The context associated with any transient handle in the command will be flushed when this command completes.
     *  CLEAR (0): No context is flushed as a side effect of this command.
     */
    TPMA_CC[TPMA_CC["flushed"] = 16777216] = "flushed";
    /**
     *  indicates the number of the handles in the handle area for this command
     */
    TPMA_CC[TPMA_CC["cHandles_BIT_MASK"] = 234881024] = "cHandles_BIT_MASK";
    TPMA_CC[TPMA_CC["cHandles_BIT_OFFSET"] = 25] = "cHandles_BIT_OFFSET";
    TPMA_CC[TPMA_CC["cHandles_BIT_LENGTH"] = 3] = "cHandles_BIT_LENGTH";
    TPMA_CC[TPMA_CC["cHandles_BIT_0"] = 33554432] = "cHandles_BIT_0";
    TPMA_CC[TPMA_CC["cHandles_BIT_1"] = 67108864] = "cHandles_BIT_1";
    TPMA_CC[TPMA_CC["cHandles_BIT_2"] = 134217728] = "cHandles_BIT_2";
    /**
     *  SET (1): indicates the presence of the handle area in the response
     */
    TPMA_CC[TPMA_CC["rHandle"] = 268435456] = "rHandle";
    /**
     *  SET (1): indicates that the command is vendor-specific
     *  CLEAR (0): indicates that the command is defined in a version of this specification
     */
    TPMA_CC[TPMA_CC["V"] = 536870912] = "V";
    /**
     *  allocated for software; shall be zero
     */
    TPMA_CC[TPMA_CC["Res_BIT_MASK"] = 3221225472] = "Res_BIT_MASK";
    TPMA_CC[TPMA_CC["Res_BIT_OFFSET"] = 30] = "Res_BIT_OFFSET";
    TPMA_CC[TPMA_CC["Res_BIT_LENGTH"] = 2] = "Res_BIT_LENGTH";
    TPMA_CC[TPMA_CC["Res_BIT_0"] = 1073741824] = "Res_BIT_0";
    TPMA_CC[TPMA_CC["Res_BIT_1"] = 2147483648] = "Res_BIT_1";
})(TPMA_CC = exports.TPMA_CC // UINT32
 || (exports.TPMA_CC // UINT32
 = {}));
; // enum TPMA_CC
/**
 *  This structure of this attribute is used to report that the TPM is designed for these modes. This structure may be read using TPM2_GetCapability(capability = TPM_CAP_TPM_PROPERTIES, property = TPM_PT_MODES).
 */
var TPMA_MODES;
(function (TPMA_MODES) {
    /**
     *  SET (1): indicates that the TPM is designed to comply with all of the FIPS 140-2 requirements at Level 1 or higher.
     */
    TPMA_MODES[TPMA_MODES["FIPS_140_2"] = 1] = "FIPS_140_2";
})(TPMA_MODES = exports.TPMA_MODES // UINT32
 || (exports.TPMA_MODES // UINT32
 = {}));
; // TPMA_MODES
/**
 *  A TPM_NV_INDEX is used to reference a defined location in NV memory. The format of the Index is changed from TPM 1.2 in order to include the Index in the reserved handle space. Handles in this range use the digest of the public area of the Index as the Name of the entity in authorization computations
 */
var TPM_NV_INDEX;
(function (TPM_NV_INDEX) {
    /**
     *  The Index of the NV location
     */
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_MASK"] = 16777215] = "index_BIT_MASK";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_OFFSET"] = 0] = "index_BIT_OFFSET";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_LENGTH"] = 24] = "index_BIT_LENGTH";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_0"] = 1] = "index_BIT_0";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_1"] = 2] = "index_BIT_1";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_2"] = 4] = "index_BIT_2";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_3"] = 8] = "index_BIT_3";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_4"] = 16] = "index_BIT_4";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_5"] = 32] = "index_BIT_5";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_6"] = 64] = "index_BIT_6";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_7"] = 128] = "index_BIT_7";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_8"] = 256] = "index_BIT_8";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_9"] = 512] = "index_BIT_9";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_10"] = 1024] = "index_BIT_10";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_11"] = 2048] = "index_BIT_11";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_12"] = 4096] = "index_BIT_12";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_13"] = 8192] = "index_BIT_13";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_14"] = 16384] = "index_BIT_14";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_15"] = 32768] = "index_BIT_15";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_16"] = 65536] = "index_BIT_16";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_17"] = 131072] = "index_BIT_17";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_18"] = 262144] = "index_BIT_18";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_19"] = 524288] = "index_BIT_19";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_20"] = 1048576] = "index_BIT_20";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_21"] = 2097152] = "index_BIT_21";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_22"] = 4194304] = "index_BIT_22";
    TPM_NV_INDEX[TPM_NV_INDEX["index_BIT_23"] = 8388608] = "index_BIT_23";
    /**
     *  constant value of TPM_HT_NV_INDEX indicating the NV Index range
     */
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_MASK"] = 4278190080] = "RhNv_BIT_MASK";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_OFFSET"] = 24] = "RhNv_BIT_OFFSET";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_LENGTH"] = 8] = "RhNv_BIT_LENGTH";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_0"] = 16777216] = "RhNv_BIT_0";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_1"] = 33554432] = "RhNv_BIT_1";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_2"] = 67108864] = "RhNv_BIT_2";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_3"] = 134217728] = "RhNv_BIT_3";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_4"] = 268435456] = "RhNv_BIT_4";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_5"] = 536870912] = "RhNv_BIT_5";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_6"] = 1073741824] = "RhNv_BIT_6";
    TPM_NV_INDEX[TPM_NV_INDEX["RhNv_BIT_7"] = 2147483648] = "RhNv_BIT_7";
})(TPM_NV_INDEX = exports.TPM_NV_INDEX // UINT32
 || (exports.TPM_NV_INDEX // UINT32
 = {}));
; // enum TPM_NV_INDEX
/**
 *  This structure allows the TPM to keep track of the data and permissions to manipulate an NV Index.
 */
var TPMA_NV;
(function (TPMA_NV) {
    /**
     *  SET (1): The Index data can be written if Platform Authorization is provided.
     *  CLEAR (0): Writing of the Index data cannot be authorized with Platform Authorization.
     */
    TPMA_NV[TPMA_NV["PPWRITE"] = 1] = "PPWRITE";
    /**
     *  SET (1): The Index data can be written if Owner Authorization is provided.
     *  CLEAR (0): Writing of the Index data cannot be authorized with Owner Authorization.
     */
    TPMA_NV[TPMA_NV["OWNERWRITE"] = 2] = "OWNERWRITE";
    /**
     *  SET (1): Authorizations to change the Index contents that require USER role may be provided with an HMAC session or password.
     *  CLEAR (0): Authorizations to change the Index contents that require USER role may not be provided with an HMAC session or password.
     */
    TPMA_NV[TPMA_NV["AUTHWRITE"] = 4] = "AUTHWRITE";
    /**
     *  SET (1): Authorizations to change the Index contents that require USER role may be provided with a policy session.
     *  CLEAR (0): Authorizations to change the Index contents that require USER role may not be provided with a policy session.
     *  NOTE	TPM2_NV_ChangeAuth() always requires that authorization be provided in a policy session.
     */
    TPMA_NV[TPMA_NV["POLICYWRITE"] = 8] = "POLICYWRITE";
    /**
     *  Ordinary  contains data that is opaque to the TPM that can only be modified using TPM2_NV_Write().
     */
    TPMA_NV[TPMA_NV["ORDINARY"] = 0] = "ORDINARY";
    /**
     *  Counter  contains an 8-octet value that is to be used as a counter and can only be modified with TPM2_NV_Increment()
     */
    TPMA_NV[TPMA_NV["COUNTER"] = 16] = "COUNTER";
    /**
     *  Bit Field  contains an 8-octet value to be used as a bit field and can only be modified with TPM2_NV_SetBits().
     */
    TPMA_NV[TPMA_NV["BITS"] = 32] = "BITS";
    /**
     *  Extend  contains a digest-sized value used like a PCR. The Index can only be modified using TPM2_NV_Extend(). The extend will use the nameAlg of the Index.
     */
    TPMA_NV[TPMA_NV["EXTEND"] = 64] = "EXTEND";
    /**
     *  PIN Fail - contains pinCount that increments on a PIN authorization failure and a pinLimit
     */
    TPMA_NV[TPMA_NV["PIN_FAIL"] = 128] = "PIN_FAIL";
    /**
     *  PIN Pass - contains pinCount that increments on a PIN authorization success and a pinLimit
     */
    TPMA_NV[TPMA_NV["PIN_PASS"] = 144] = "PIN_PASS";
    /**
     *  The type of the index. NOTE A TPM is not required to support all TPM_NT values
     */
    TPMA_NV[TPMA_NV["TpmNt_BIT_MASK"] = 240] = "TpmNt_BIT_MASK";
    TPMA_NV[TPMA_NV["TpmNt_BIT_OFFSET"] = 4] = "TpmNt_BIT_OFFSET";
    TPMA_NV[TPMA_NV["TpmNt_BIT_LENGTH"] = 4] = "TpmNt_BIT_LENGTH";
    TPMA_NV[TPMA_NV["TpmNt_BIT_0"] = 16] = "TpmNt_BIT_0";
    TPMA_NV[TPMA_NV["TpmNt_BIT_1"] = 32] = "TpmNt_BIT_1";
    TPMA_NV[TPMA_NV["TpmNt_BIT_2"] = 64] = "TpmNt_BIT_2";
    TPMA_NV[TPMA_NV["TpmNt_BIT_3"] = 128] = "TpmNt_BIT_3";
    /**
     *  SET (1): Index may not be deleted unless the authPolicy is satisfied using TPM2_NV_UndefineSpaceSpecial().
     *  CLEAR (0): Index may be deleted with proper platform or owner authorization using TPM2_NV_UndefineSpace(). NOTE An Index with this attribute and a policy that cannot be satisfied (e.g., an Empty Policy) cannot be deleted.
     */
    TPMA_NV[TPMA_NV["POLICY_DELETE"] = 1024] = "POLICY_DELETE";
    /**
     *  SET (1): Index cannot be written.
     *  CLEAR (0): Index can be written.
     */
    TPMA_NV[TPMA_NV["WRITELOCKED"] = 2048] = "WRITELOCKED";
    /**
     *  SET (1): A partial write of the Index data is not allowed. The write size shall match the defined space size.
     *  CLEAR (0): Partial writes are allowed. This setting is required if the .dataSize of the Index is larger than NV_MAX_BUFFER_SIZE for the implementation.
     */
    TPMA_NV[TPMA_NV["WRITEALL"] = 4096] = "WRITEALL";
    /**
     *  SET (1): TPM2_NV_WriteLock() may be used to prevent further writes to this location.
     *  CLEAR (0): TPM2_NV_WriteLock() does not block subsequent writes if TPMA_NV_WRITE_STCLEAR is also CLEAR.
     */
    TPMA_NV[TPMA_NV["WRITEDEFINE"] = 8192] = "WRITEDEFINE";
    /**
     *  SET (1): TPM2_NV_WriteLock() may be used to prevent further writes to this location until the next TPM Reset or TPM Restart.
     *  CLEAR (0): TPM2_NV_WriteLock() does not block subsequent writes if TPMA_NV_WRITEDEFINE is also CLEAR.
     */
    TPMA_NV[TPMA_NV["WRITE_STCLEAR"] = 16384] = "WRITE_STCLEAR";
    /**
     *  SET (1): If TPM2_NV_GlobalWriteLock() is successful, then further writes to this location are not permitted until the next TPM Reset or TPM Restart.
     *  CLEAR (0): TPM2_NV_GlobalWriteLock() has no effect on the writing of the data at this Index.
     */
    TPMA_NV[TPMA_NV["GLOBALLOCK"] = 32768] = "GLOBALLOCK";
    /**
     *  SET (1): The Index data can be read if Platform Authorization is provided.
     *  CLEAR (0): Reading of the Index data cannot be authorized with Platform Authorization.
     */
    TPMA_NV[TPMA_NV["PPREAD"] = 65536] = "PPREAD";
    /**
     *  SET (1): The Index data can be read if Owner Authorization is provided.
     *  CLEAR (0): Reading of the Index data cannot be authorized with Owner Authorization.
     */
    TPMA_NV[TPMA_NV["OWNERREAD"] = 131072] = "OWNERREAD";
    /**
     *  SET (1): The Index data may be read if the authValue is provided.
     *  CLEAR (0): Reading of the Index data cannot be authorized with the Index authValue.
     */
    TPMA_NV[TPMA_NV["AUTHREAD"] = 262144] = "AUTHREAD";
    /**
     *  SET (1): The Index data may be read if the authPolicy is satisfied.
     *  CLEAR (0): Reading of the Index data cannot be authorized with the Index authPolicy.
     */
    TPMA_NV[TPMA_NV["POLICYREAD"] = 524288] = "POLICYREAD";
    /**
     *  SET (1): Authorization failures of the Index do not affect the DA logic and authorization of the Index is not blocked when the TPM is in Lockout mode.
     *  CLEAR (0): Authorization failures of the Index will increment the authorization failure counter and authorizations of this Index are not allowed when the TPM is in Lockout mode.
     */
    TPMA_NV[TPMA_NV["NO_DA"] = 33554432] = "NO_DA";
    /**
     *  SET (1): NV Index state is only required to be saved when the TPM performs an orderly shutdown (TPM2_Shutdown()).
     *  CLEAR (0): NV Index state is required to be persistent after the command to update the Index completes successfully (that is, the NV update is synchronous with the update command). NOTE If TPMA_NV_ORDERLY is SET, TPMA_NV_WRITTEN will be CLEAR by TPM Reset.
     */
    TPMA_NV[TPMA_NV["ORDERLY"] = 67108864] = "ORDERLY";
    /**
     *  SET (1): TPMA_NV_WRITTEN for the Index is CLEAR by TPM Reset or TPM Restart.
     *  CLEAR (0): TPMA_NV_WRITTEN is not changed by TPM Restart.
     *  NOTE	This attribute may only be SET if TPM_NT is not TPM_NT_COUNTER.
     */
    TPMA_NV[TPMA_NV["CLEAR_STCLEAR"] = 134217728] = "CLEAR_STCLEAR";
    /**
     *  SET (1): Reads of the Index are blocked until the next TPM Reset or TPM Restart.
     *  CLEAR (0): Reads of the Index are allowed if proper authorization is provided.
     */
    TPMA_NV[TPMA_NV["READLOCKED"] = 268435456] = "READLOCKED";
    /**
     *  SET (1): Index has been written.
     *  CLEAR (0): Index has not been written.
     */
    TPMA_NV[TPMA_NV["WRITTEN"] = 536870912] = "WRITTEN";
    /**
     *  SET (1): This Index may be undefined with Platform Authorization but not with Owner Authorization.
     *  CLEAR (0): This Index may be undefined using Owner Authorization but not with Platform Authorization. The TPM will validate that this attribute is SET when the Index is defined using Platform Authorization and will validate that this attribute is CLEAR when the Index is defined using Owner Authorization.
     */
    TPMA_NV[TPMA_NV["PLATFORMCREATE"] = 1073741824] = "PLATFORMCREATE";
    /**
     *  SET (1): TPM2_NV_ReadLock() may be used to SET TPMA_NV_READLOCKED for this Index.
     *  CLEAR (0): TPM2_NV_ReadLock() has no effect on this Index.
     */
    TPMA_NV[TPMA_NV["READ_STCLEAR"] = 2147483648] = "READ_STCLEAR";
})(TPMA_NV = exports.TPMA_NV // UINT32
 || (exports.TPMA_NV // UINT32
 = {}));
; // enum TPMA_NV
// Equivalent of UnionElementFromSelector in TSS.Net
function createUnion(unionType, selector) {
    var u = null;
    switch (unionType) {
        case 'TPMU_PUBLIC_PARMS':
            switch (selector) {
                case TPM_ALG_ID.KEYEDHASH:
                    u = new TPMS_KEYEDHASH_PARMS();
                    break;
                //case TPM_ALG_ID.SYMCIPHER: u = new TPMS_SYMCIPHER_PARMS(); break;
                case TPM_ALG_ID.RSA:
                    u = new TPMS_RSA_PARMS();
                    break;
            }
            break;
        case 'TPMU_PUBLIC_ID':
            switch (selector) {
                case TPM_ALG_ID.KEYEDHASH:
                    u = new TPM2B_DIGEST_Keyedhash();
                    break;
                //case TPM_ALG_ID.SYMCIPHER: u = new TPM2B_DIGEST_Symcipher(); break;
                case TPM_ALG_ID.RSA:
                    u = new TPM2B_PUBLIC_KEY_RSA();
                    break;
            }
            break;
        case 'TPMU_ASYM_SCHEME':
            switch (selector) {
                //case TPM_ALG_ID.ECDH: u = new TPMS_KEY_SCHEME_ECDH(); break;
                //case TPM_ALG_ID.ECMQV: u = new TPMS_KEY_SCHEME_ECMQV(); break;
                case TPM_ALG_ID.RSASSA:
                    u = new TPMS_SIG_SCHEME_RSASSA();
                    break;
                //case TPM_ALG_ID.RSAPSS: u = new TPMS_SIG_SCHEME_RSAPSS(); break;
                //case TPM_ALG_ID.ECDSA: u = new TPMS_SIG_SCHEME_ECDSA(); break;
                //case TPM_ALG_ID.ECDAA: u = new TPMS_SIG_SCHEME_ECDAA(); break;
                // code generator workaround BUGBUG >> (probChild)case TPM_ALG_ID.SM2: u = new TPMS_SIG_SCHEME_SM2(); break;
                // code generator workaround BUGBUG >> (probChild)case TPM_ALG_ID.ECSCHNORR: u = new TPMS_SIG_SCHEME_ECSCHNORR(); break;
                //case TPM_ALG_ID.RSAES: u = new TPMS_ENC_SCHEME_RSAES(); break;
                case TPM_ALG_ID.OAEP:
                    u = new TPMS_ENC_SCHEME_OAEP();
                    break;
                //case TPM_ALG_ID.ANY: u = new TPMS_SCHEME_HASH(); break;
                case TPM_ALG_ID.NULL:
                    u = new TPMS_NULL_ASYM_SCHEME();
                    break;
            }
            break;
        case 'TPMU_SCHEME_KEYEDHASH':
            switch (selector) {
                case TPM_ALG_ID.HMAC:
                    u = new TPMS_SCHEME_HMAC();
                    break;
                case TPM_ALG_ID.XOR:
                    u = new TPMS_SCHEME_XOR();
                    break;
                case TPM_ALG_ID.NULL:
                    u = new TPMS_NULL_SCHEME_KEYEDHASH();
                    break;
            }
            break;
        case 'TPMU_CAPABILITIES':
            switch (selector) {
                case TPM_CAP.TPM_PROPERTIES:
                    u = new TPML_TAGGED_TPM_PROPERTY();
                    break;
            }
            break;
        default:
            throw (new Error('CreateUnion(' + unionType + ', ' + TPM_ALG_ID[selector] + '): Unrecognized union type'));
    }
    if (u == null)
        throw (new Error('CreateUnion(' + unionType + ', ' + TPM_ALG_ID[selector] + '): Unrecognized selector'));
    return u;
} // CreateUnion<U>()
/**
* This structure provides information relating to the creation environment for the object. The creation data includes the parent Name, parent Qualified Name, and the digest of selected PCR. These values represent the environment in which the object was created. Creation data allows a relying party to determine if an object was created when some appropriate protections were present.
*/
var TPMS_CREATION_DATA = (function (_super) {
    __extends(TPMS_CREATION_DATA, _super);
    /**
    * @param _pcrSelect list indicating the PCR included in pcrDigest
    * @param _pcrDigest digest of the selected PCR using nameAlg of the object for which this structure is being created pcrDigest.size shall be zero if the pcrSelect list is empty.
    * @param _locality the locality at which the object was created
    * @param _parentNameAlg nameAlg of the parent
    * @param _parentName Name of the parent at time of creation The size will match digest size associated with parentNameAlg unless it is TPM_ALG_NULL, in which case the size will be 4 and parentName will be the hierarchy handle.
    * @param _parentQualifiedName Qualified Name of the parent at the time of creation Size is the same as parentName.
    * @param _outsideInfo association with additional information added by the key creator This will be the contents of the outsideInfo parameter in TPM2_Create() or TPM2_CreatePrimary().
    */
    function TPMS_CREATION_DATA(pcrSelect, pcrDigest, locality, parentNameAlg, parentName, parentQualifiedName, outsideInfo) {
        if (pcrSelect === void 0) { pcrSelect = null; }
        if (pcrDigest === void 0) { pcrDigest = null; }
        if (locality === void 0) { locality = 0; }
        if (parentNameAlg === void 0) { parentNameAlg = 0; }
        if (parentName === void 0) { parentName = null; }
        if (parentQualifiedName === void 0) { parentQualifiedName = null; }
        if (outsideInfo === void 0) { outsideInfo = null; }
        var _this = _super.call(this) || this;
        _this.pcrSelect = pcrSelect;
        _this.pcrDigest = pcrDigest;
        _this.locality = locality;
        _this.parentNameAlg = parentNameAlg;
        _this.parentName = parentName;
        _this.parentQualifiedName = parentQualifiedName;
        _this.outsideInfo = outsideInfo;
        return _this;
    }
    /** TpmMarshaller method */
    TPMS_CREATION_DATA.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.arrayToTpm(this.pcrSelect, buf, 4, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.pcrDigest, buf, pos);
        pos = TpmMarshaller_js_1.toTpm(this.locality, buf, 1, pos);
        pos = TpmMarshaller_js_1.toTpm(this.parentNameAlg, buf, 2, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.parentName, buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.parentQualifiedName, buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.outsideInfo, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_CREATION_DATA.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.arrayFromTpm(TPMS_PCR_SELECTION, buf, 4, pos), this.pcrSelect = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.pcrDigest = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm(buf, 1, pos), this.locality = _c[0], pos = _c[1];
        _d = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.parentNameAlg = _d[0], pos = _d[1];
        _e = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.parentName = _e[0], pos = _e[1];
        _f = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.parentQualifiedName = _f[0], pos = _f[1];
        _g = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.outsideInfo = _g[0], pos = _g[1];
        return pos;
        var _a, _b, _c, _d, _e, _f, _g;
    };
    return TPMS_CREATION_DATA;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMS_CREATION_DATA = TPMS_CREATION_DATA;
; // class TPMS_CREATION_DATA
/**
* This structure is used for sizing the TPM2B_ID_OBJECT.
*/
var TPMS_ID_OBJECT = (function (_super) {
    __extends(TPMS_ID_OBJECT, _super);
    function TPMS_ID_OBJECT(
        /**
        * HMAC using the nameAlg of the storage key on the target TPM
        */
        integrityHMAC, 
        /**
        * credential protector information returned if name matches the referenced object All of the encIdentity is encrypted, including the size field. NOTE The TPM is not required to check that the size is not larger than the digest of the nameAlg. However, if the size is larger, the ID object may not be usable on a TPM that has no digest larger than produced by nameAlg.
        */
        encIdentity) {
        if (integrityHMAC === void 0) { integrityHMAC = null; }
        if (encIdentity === void 0) { encIdentity = null; }
        var _this = _super.call(this) || this;
        _this.integrityHMAC = integrityHMAC;
        _this.encIdentity = encIdentity;
        return _this;
    }
    /** TpmMarshaller method */
    TPMS_ID_OBJECT.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.integrityHMAC, buf, pos);
        this.encIdentity.copy(buf, pos);
        pos += this.encIdentity.length;
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_ID_OBJECT.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.integrityHMAC = _a[0], pos = _a[1];
        var size = TpmMarshaller_js_1.getCurStuctRemainingSize(pos);
        this.encIdentity = new Buffer(size);
        buf.copy(this.encIdentity, 0, pos, pos + size);
        pos += size;
        return pos;
        var _a;
    };
    return TPMS_ID_OBJECT;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMS_ID_OBJECT = TPMS_ID_OBJECT;
; // class TPMS_ID_OBJECT
/**
 *  These are the RSA schemes that only need a hash algorithm as a scheme parameter.
 */
var TPMS_ENC_SCHEME_OAEP = (function () {
    function TPMS_ENC_SCHEME_OAEP(
        /**
         *  the hash algorithm used to digest the message
         */
        hashAlg) {
        if (hashAlg === void 0) { hashAlg = 0; }
        this.hashAlg = hashAlg;
    }
    /** TpmUnion method */
    TPMS_ENC_SCHEME_OAEP.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.OAEP;
    };
    /** TpmMarshaller method */
    TPMS_ENC_SCHEME_OAEP.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.hashAlg, buf, 2, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_ENC_SCHEME_OAEP.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.hashAlg = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPMS_ENC_SCHEME_OAEP;
}());
exports.TPMS_ENC_SCHEME_OAEP = TPMS_ENC_SCHEME_OAEP;
; // class TPMS_ENC_SCHEME_OAEP
/**
 *  This structure describes the parameters that would appear in the public area of a KEYEDHASH object.
 */
var TPMS_KEYEDHASH_PARMS = (function () {
    /**
    * @param _scheme Indicates the signing method used for a keyedHash signing object. This field also determines the size of the data field for a data object created with TPM2_Create() or TPM2_CreatePrimary(). (One of TPMS_SCHEME_HMAC, TPMS_SCHEME_XOR, TPMS_NULL_SCHEME_KEYEDHASH)
    */
    function TPMS_KEYEDHASH_PARMS(
        /**
         *  the hash algorithm used to digest the message
         */
        scheme) {
        if (scheme === void 0) { scheme = null; }
        this.scheme = scheme;
    }
    /** TpmUnion method */
    TPMS_KEYEDHASH_PARMS.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.KEYEDHASH;
    };
    /** TpmMarshaller method */
    TPMS_KEYEDHASH_PARMS.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.scheme.GetUnionSelector(), buf, 2, pos);
        pos = this.scheme.toTpm(buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_KEYEDHASH_PARMS.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        var schemeAlg;
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), schemeAlg = _a[0], pos = _a[1];
        this.scheme = createUnion('TPMU_SCHEME_KEYEDHASH', schemeAlg);
        pos = this.scheme.fromTpm(buf, pos);
        return pos;
        var _a;
    };
    return TPMS_KEYEDHASH_PARMS;
}());
exports.TPMS_KEYEDHASH_PARMS = TPMS_KEYEDHASH_PARMS;
; // class TPMS_KEYEDHASH_PARMS
/**
 *  Custom data structure representing an empty element (i.e. the one with no data to marshal) for selector algorithm TPM_ALG_NULL for the union TpmuAsymScheme
 */
var TPMS_NULL_ASYM_SCHEME = (function () {
    function TPMS_NULL_ASYM_SCHEME() {
    }
    /** TpmUnion method */
    TPMS_NULL_ASYM_SCHEME.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.NULL;
    };
    /** TpmMarshaller method */
    TPMS_NULL_ASYM_SCHEME.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_NULL_ASYM_SCHEME.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        return pos;
    };
    return TPMS_NULL_ASYM_SCHEME;
}());
exports.TPMS_NULL_ASYM_SCHEME = TPMS_NULL_ASYM_SCHEME;
; // class TPMS_NULL_ASYM_SCHEME
/**
 *  Table 145 Definition of Types for HMAC_SIG_SCHEME
 */
var TPMS_NULL_SCHEME_KEYEDHASH = (function () {
    function TPMS_NULL_SCHEME_KEYEDHASH() {
    }
    /** TpmUnion method */
    TPMS_NULL_SCHEME_KEYEDHASH.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.NULL;
    };
    /** TpmMarshaller method */
    TPMS_NULL_SCHEME_KEYEDHASH.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_NULL_SCHEME_KEYEDHASH.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        return pos;
    };
    return TPMS_NULL_SCHEME_KEYEDHASH;
}());
exports.TPMS_NULL_SCHEME_KEYEDHASH = TPMS_NULL_SCHEME_KEYEDHASH;
; // class TPMS_NULL_SCHEME_KEYEDHASH
/**
 * Table 87 Definition of TPMS_PCR_SELECTION Structure
 */
var TPMS_PCR_SELECTION = (function (_super) {
    __extends(TPMS_PCR_SELECTION, _super);
    function TPMS_PCR_SELECTION(
        /**
        * the hash algorithm associated with the selection
        */
        hash, 
        /**
        * the bit map of selected PCR
        */
        pcrSelect) {
        if (hash === void 0) { hash = 0; }
        if (pcrSelect === void 0) { pcrSelect = null; }
        var _this = _super.call(this) || this;
        _this.hash = hash;
        _this.pcrSelect = pcrSelect;
        return _this;
    }
    /** TpmMarshaller method */
    TPMS_PCR_SELECTION.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.hash, buf, 2, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.pcrSelect, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_PCR_SELECTION.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.hash = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.pcrSelect = _b[0], pos = _b[1];
        return pos;
        var _a, _b;
    };
    return TPMS_PCR_SELECTION;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMS_PCR_SELECTION = TPMS_PCR_SELECTION;
; // class TPMS_PCR_SELECTION
/**
 *  A TPM compatible with this specification and supporting RSA shall support two primes and an exponent of zero. Support for other values is optional. Use of other exponents in duplicated keys is not recommended because the resulting keys would not be interoperable with other TPMs.
 */
var TPMS_RSA_PARMS = (function () {
    /**
    */
    function TPMS_RSA_PARMS(
        /**
         *  @param _symmetric for a restricted decryption key, shall be set to a supported symmetric algorithm, key size, and mode. if the key is not a restricted decryption key, this field shall be set to TPM_ALG_NULL.
         */
        symmetric, 
        /**
         *  @param _scheme scheme.scheme shall be: for an unrestricted signing key, either TPM_ALG_RSAPSS TPM_ALG_RSASSA or TPM_ALG_NULL for a restricted signing key, either TPM_ALG_RSAPSS or TPM_ALG_RSASSA for an unrestricted decryption key, TPM_ALG_RSAES, TPM_ALG_OAEP, or TPM_ALG_NULL unless the object also has the sign attribute for a restricted decryption key, TPM_ALG_NULL NOTE When both sign and decrypt are SET, restricted shall be CLEAR and scheme shall be TPM_ALG_NULL. (One of TPMS_KEY_SCHEME_ECDH, TPMS_KEY_SCHEME_ECMQV, TPMS_SIG_SCHEME_RSASSA, TPMS_SIG_SCHEME_RSAPSS, TPMS_SIG_SCHEME_ECDSA, TPMS_SIG_SCHEME_ECDAA, TPMS_SIG_SCHEME_SM2, TPMS_SIG_SCHEME_ECSCHNORR, TPMS_ENC_SCHEME_RSAES, TPMS_ENC_SCHEME_OAEP, TPMS_SCHEME_HASH, TPMS_NULL_ASYM_SCHEME)
         */
        scheme, 
        /**
         *  @param _keyBits number of bits in the public modulus
         */
        keyBits, 
        /**
         *  @param _exponent the public exponent A prime number greater than 2. When zero, indicates that the exponent is the default of 216 + 1
         */
        exponent) {
        if (symmetric === void 0) { symmetric = null; }
        if (scheme === void 0) { scheme = null; }
        if (keyBits === void 0) { keyBits = 0; }
        if (exponent === void 0) { exponent = 0; }
        this.symmetric = symmetric;
        this.scheme = scheme;
        this.keyBits = keyBits;
        this.exponent = exponent;
    }
    /** TpmUnion method */
    TPMS_RSA_PARMS.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.RSA;
    };
    /** TpmMarshaller method */
    TPMS_RSA_PARMS.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = this.symmetric.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm(this.scheme.GetUnionSelector(), buf, 2, pos);
        pos = this.scheme.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm(this.keyBits, buf, 2, pos);
        pos = TpmMarshaller_js_1.toTpm(this.exponent, buf, 4, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_RSA_PARMS.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.createFromTpm(TPMT_SYM_DEF_OBJECT, buf, pos), this.symmetric = _a[0], pos = _a[1];
        var schemeAlg;
        _b = TpmMarshaller_js_1.fromTpm(buf, 2, pos), schemeAlg = _b[0], pos = _b[1];
        this.scheme = createUnion('TPMU_ASYM_SCHEME', schemeAlg);
        pos = this.scheme.fromTpm(buf, pos);
        _c = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.keyBits = _c[0], pos = _c[1];
        _d = TpmMarshaller_js_1.fromTpm(buf, 4, pos), this.exponent = _d[0], pos = _d[1];
        return pos;
        var _a, _b, _c, _d;
    };
    return TPMS_RSA_PARMS;
}());
exports.TPMS_RSA_PARMS = TPMS_RSA_PARMS;
; // class TPMS_RSA_PARMS
/**
 *  Table 145 Definition of Types for HMAC_SIG_SCHEME
 */
var TPMS_SCHEME_HMAC = (function () {
    function TPMS_SCHEME_HMAC(
        /**
         *  the hash algorithm used to digest the message
         */
        hashAlg) {
        if (hashAlg === void 0) { hashAlg = 0; }
        this.hashAlg = hashAlg;
    }
    /** TpmUnion method */
    TPMS_SCHEME_HMAC.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.HMAC;
    };
    /** TpmMarshaller method */
    TPMS_SCHEME_HMAC.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.hashAlg, buf, 2, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_SCHEME_HMAC.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.hashAlg = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPMS_SCHEME_HMAC;
}());
exports.TPMS_SCHEME_HMAC = TPMS_SCHEME_HMAC;
; // class TPMS_SCHEME_HMAC
/**
 *  This structure is for the XOR encryption scheme.
 */
var TPMS_SCHEME_XOR = (function () {
    function TPMS_SCHEME_XOR(
        /**
         *  the hash algorithm used to digest the message
         */
        hashAlg, kdf) {
        if (hashAlg === void 0) { hashAlg = 0; }
        if (kdf === void 0) { kdf = 0; }
        this.hashAlg = hashAlg;
        this.kdf = kdf;
    }
    /** TpmUnion method */
    TPMS_SCHEME_XOR.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.XOR;
    };
    /** TpmMarshaller method */
    TPMS_SCHEME_XOR.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.hashAlg, buf, 2, pos);
        pos = TpmMarshaller_js_1.toTpm(this.kdf, buf, 2, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_SCHEME_XOR.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.hashAlg = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.kdf = _b[0], pos = _b[1];
        return pos;
        var _a, _b;
    };
    return TPMS_SCHEME_XOR;
}());
exports.TPMS_SCHEME_XOR = TPMS_SCHEME_XOR;
; // class TPMS_SCHEME_XOR
/**
 * This structure defines the values to be placed in the sensitive area of a created object. This structure is only used within a TPM2B_SENSITIVE_CREATE structure.
 */
var TPMS_SENSITIVE_CREATE = (function (_super) {
    __extends(TPMS_SENSITIVE_CREATE, _super);
    function TPMS_SENSITIVE_CREATE(
        /**
         * the USER auth secret value
         */
        userAuth, 
        /**
         *  data to be sealed, a key, or derivation values
         */
        data) {
        if (userAuth === void 0) { userAuth = null; }
        if (data === void 0) { data = null; }
        var _this = _super.call(this) || this;
        _this.userAuth = userAuth;
        _this.data = data;
        return _this;
    }
    /** TpmMarshaller method */
    TPMS_SENSITIVE_CREATE.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.userAuth, buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.data, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_SENSITIVE_CREATE.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.userAuth = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.data = _b[0], pos = _b[1];
        return pos;
        var _a, _b;
    };
    return TPMS_SENSITIVE_CREATE;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMS_SENSITIVE_CREATE = TPMS_SENSITIVE_CREATE;
; // class TPMS_SENSITIVE_CREATE
/**
 *  These are the RSA schemes that only need a hash algorithm as a scheme parameter.
 */
var TPMS_SIG_SCHEME_RSASSA = (function () {
    function TPMS_SIG_SCHEME_RSASSA(
        /**
         *  the hash algorithm used to digest the message
         */
        hashAlg) {
        if (hashAlg === void 0) { hashAlg = 0; }
        this.hashAlg = hashAlg;
    }
    /** TpmUnion method */
    TPMS_SIG_SCHEME_RSASSA.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.RSASSA;
    };
    /** TpmMarshaller method */
    TPMS_SIG_SCHEME_RSASSA.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.hashAlg, buf, 2, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_SIG_SCHEME_RSASSA.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.hashAlg = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPMS_SIG_SCHEME_RSASSA;
}());
exports.TPMS_SIG_SCHEME_RSASSA = TPMS_SIG_SCHEME_RSASSA;
; // class TPMS_SIG_SCHEME_RSASSA
/**
 * This structure is used to report the properties that are UINT32 values. It is returned in response to a TPM2_GetCapability().
 */
var TPMS_TAGGED_PROPERTY = (function (_super) {
    __extends(TPMS_TAGGED_PROPERTY, _super);
    function TPMS_TAGGED_PROPERTY(
        /**
         * a property identifier
         */
        property, 
        /**
         * the value of the property
         */
        value) {
        if (property === void 0) { property = 0; }
        if (value === void 0) { value = 0; }
        var _this = _super.call(this) || this;
        _this.property = property;
        _this.value = value;
        return _this;
    }
    /** TpmMarshaller method */
    TPMS_TAGGED_PROPERTY.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.property, buf, 4, pos);
        pos = TpmMarshaller_js_1.toTpm(this.value, buf, 4, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMS_TAGGED_PROPERTY.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 4, pos), this.property = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm(buf, 4, pos), this.value = _b[0], pos = _b[1];
        return pos;
        var _a, _b;
    };
    return TPMS_TAGGED_PROPERTY;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMS_TAGGED_PROPERTY = TPMS_TAGGED_PROPERTY;
; // class TPMS_TAGGED_PROPERTY
/**
 * This list is used to report on a list of properties that are TPMS_TAGGED_PROPERTY values. It is returned by a TPM2_GetCapability().
 */
var TPML_TAGGED_TPM_PROPERTY = (function (_super) {
    __extends(TPML_TAGGED_TPM_PROPERTY, _super);
    function TPML_TAGGED_TPM_PROPERTY(
        /**
         * an array of tagged properties
         */
        tpmProperty) {
        if (tpmProperty === void 0) { tpmProperty = null; }
        var _this = _super.call(this) || this;
        _this.tpmProperty = tpmProperty;
        return _this;
    }
    /** TpmUnion method */
    TPML_TAGGED_TPM_PROPERTY.prototype.GetUnionSelector = function () {
        return TPM_CAP.TPM_PROPERTIES;
    };
    /** TpmMarshaller method */
    TPML_TAGGED_TPM_PROPERTY.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.arrayToTpm(this.tpmProperty, buf, 4, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPML_TAGGED_TPM_PROPERTY.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.arrayFromTpm(TPMS_TAGGED_PROPERTY, buf, 4, pos), this.tpmProperty = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPML_TAGGED_TPM_PROPERTY;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPML_TAGGED_TPM_PROPERTY = TPML_TAGGED_TPM_PROPERTY;
; // class TPML_TAGGED_TPM_PROPERTY
/**
  * This structure is used for a data buffer that is required to be no larger than the size of the Name of an object.
  */
var TPM2B_DATA = (function (_super) {
    __extends(TPM2B_DATA, _super);
    function TPM2B_DATA(
        /**
         * an encrypted private area
         */
        buffer) {
        if (buffer === void 0) { buffer = null; }
        var _this = _super.call(this) || this;
        _this.buffer = buffer;
        return _this;
    }
    /** TpmMarshaller method */
    TPM2B_DATA.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.buffer, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPM2B_DATA.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.buffer = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM2B_DATA;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPM2B_DATA = TPM2B_DATA;
; // TPM2B_DATA
/**
  * This structure is used for a sized buffer that cannot be larger than the largest digest produced by any hash algorithm implemented on the TPM.
  */
var TPM2B_DIGEST = (function (_super) {
    __extends(TPM2B_DIGEST, _super);
    function TPM2B_DIGEST(
        /**
         * an encrypted private area
         */
        buffer) {
        if (buffer === void 0) { buffer = null; }
        var _this = _super.call(this) || this;
        _this.buffer = buffer;
        return _this;
    }
    /** TpmMarshaller method */
    TPM2B_DIGEST.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.buffer, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPM2B_DIGEST.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.buffer = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM2B_DIGEST;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPM2B_DIGEST = TPM2B_DIGEST;
; // TPM2B_DIGEST
/**
 * Table 182 Definition of TPM2B_ENCRYPTED_SECRET Structure
 */
var TPM2B_ENCRYPTED_SECRET = (function (_super) {
    __extends(TPM2B_ENCRYPTED_SECRET, _super);
    function TPM2B_ENCRYPTED_SECRET(
        /**
         * secret
         */
        secret) {
        if (secret === void 0) { secret = null; }
        var _this = _super.call(this) || this;
        _this.secret = secret;
        return _this;
    }
    /** TpmMarshaller method */
    TPM2B_ENCRYPTED_SECRET.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.secret, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPM2B_ENCRYPTED_SECRET.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.secret = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM2B_ENCRYPTED_SECRET;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPM2B_ENCRYPTED_SECRET = TPM2B_ENCRYPTED_SECRET;
; // TPM2B_ENCRYPTED_SECRET
/**
 * Auto-derived from TPM2B_DIGEST to provide unique GetUnionSelector() implementation
 */
var TPM2B_DIGEST_Keyedhash = (function () {
    function TPM2B_DIGEST_Keyedhash(
        /**
         * an encrypted private area
         */
        buffer) {
        if (buffer === void 0) { buffer = null; }
        this.buffer = buffer;
    }
    /** TpmUnion method */
    TPM2B_DIGEST_Keyedhash.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.KEYEDHASH;
    };
    /** TpmMarshaller method */
    TPM2B_DIGEST_Keyedhash.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.buffer, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPM2B_DIGEST_Keyedhash.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.buffer = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM2B_DIGEST_Keyedhash;
}());
exports.TPM2B_DIGEST_Keyedhash = TPM2B_DIGEST_Keyedhash;
; // TPM2B_DIGEST_Keyedhash
/**
* Auto-derived from TPM2B_DIGEST to provide unique GetUnionSelector() implementation
*/
var TPM2B_DIGEST_Symcipher = (function () {
    function TPM2B_DIGEST_Symcipher(
        /**
         * an encrypted private area
         */
        buffer) {
        if (buffer === void 0) { buffer = null; }
        this.buffer = buffer;
    }
    /** TpmUnion method */
    TPM2B_DIGEST_Symcipher.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.KEYEDHASH;
    };
    /** TpmMarshaller method */
    TPM2B_DIGEST_Symcipher.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.buffer, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPM2B_DIGEST_Symcipher.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.buffer = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM2B_DIGEST_Symcipher;
}());
exports.TPM2B_DIGEST_Symcipher = TPM2B_DIGEST_Symcipher;
; // TPM2B_DIGEST_Symcipher
/**
 * The TPM2B_PRIVATE structure is used as a parameter in multiple commands that create, load, and modify the sensitive area of an object.
 */
var TPM2B_PRIVATE = (function (_super) {
    __extends(TPM2B_PRIVATE, _super);
    function TPM2B_PRIVATE(
        /**
        * an encrypted private area
        */
        buffer) {
        if (buffer === void 0) { buffer = null; }
        var _this = _super.call(this) || this;
        _this.buffer = buffer;
        return _this;
    }
    /** TpmMarshaller method */
    TPM2B_PRIVATE.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.buffer, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPM2B_PRIVATE.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.buffer = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM2B_PRIVATE;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPM2B_PRIVATE = TPM2B_PRIVATE;
; // TPM2B_PRIVATE
/**
 * This sized buffer holds the largest RSA public key supported by the TPM.
 */
var TPM2B_PUBLIC_KEY_RSA = (function () {
    function TPM2B_PUBLIC_KEY_RSA(buffer) {
        if (buffer === void 0) { buffer = null; }
        this.buffer = buffer;
    }
    /** TpmUnion method */
    TPM2B_PUBLIC_KEY_RSA.prototype.GetUnionSelector = function () {
        return TPM_ALG_ID.RSA;
    };
    /** TpmMarshaller method */
    TPM2B_PUBLIC_KEY_RSA.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.buffer, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPM2B_PUBLIC_KEY_RSA.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.buffer = _a[0], pos = _a[1];
        return pos;
        var _a;
    };
    return TPM2B_PUBLIC_KEY_RSA;
}());
exports.TPM2B_PUBLIC_KEY_RSA = TPM2B_PUBLIC_KEY_RSA;
; // class TPM2B_PUBLIC_KEY_RSA
/**
 *  Table 191 defines the public area structure. The Name of the object is nameAlg concatenated with the digest of this structure using nameAlg.
 */
var TPMT_PUBLIC = (function (_super) {
    __extends(TPMT_PUBLIC, _super);
    function TPMT_PUBLIC(
        // Selector algorithm implicitly associated with this object
        // private TPM_ALG_ID type;
        /**
         *  algorithm used for computing the Name of the object NOTE The "+" indicates that the instance of a TPMT_PUBLIC may have a "+" to indicate that the nameAlg may be TPM_ALG_NULL.
         */
        nameAlg, 
        /**
         *  attributes that, along with type, determine the manipulations of this object
         */
        objectAttributes, 
        /**
         *  size in octets of the buffer field; may be 0
         */
        // private short authPolicySize;
        /**
         *  optional policy for using this key The policy is computed using the nameAlg of the object. NOTE Shall be the Empty Policy if no authorization policy is present.
         */
        authPolicy, 
        /**
         *  the algorithm or structure details
         */
        parameters, 
        /**
         *  the unique identifier of the structure For an asymmetric key, this would be the public key.
         */
        unique) {
        if (nameAlg === void 0) { nameAlg = 0; }
        if (objectAttributes === void 0) { objectAttributes = 0; }
        if (authPolicy === void 0) { authPolicy = null; }
        if (parameters === void 0) { parameters = null; }
        if (unique === void 0) { unique = null; }
        var _this = _super.call(this) || this;
        _this.nameAlg = nameAlg;
        _this.objectAttributes = objectAttributes;
        _this.authPolicy = authPolicy;
        _this.parameters = parameters;
        _this.unique = unique;
        return _this;
    }
    /** TpmMarshaller method */
    TPMT_PUBLIC.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.parameters.GetUnionSelector(), buf, 2, pos);
        pos = TpmMarshaller_js_1.toTpm(this.nameAlg, buf, 2, pos);
        pos = TpmMarshaller_js_1.toTpm(this.objectAttributes, buf, 4, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.authPolicy, buf, pos);
        pos = this.parameters.toTpm(buf, pos);
        pos = this.unique.toTpm(buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMT_PUBLIC.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        var type;
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), type = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.nameAlg = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm(buf, 4, pos), this.objectAttributes = _c[0], pos = _c[1];
        _d = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.authPolicy = _d[0], pos = _d[1];
        this.parameters = createUnion('TPMU_PUBLIC_PARMS', type);
        pos = this.parameters.fromTpm(buf, pos);
        this.unique = createUnion('TPMU_PUBLIC_ID', type);
        pos = this.unique.fromTpm(buf, pos);
        return pos;
        var _a, _b, _c, _d;
    };
    return TPMT_PUBLIC;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMT_PUBLIC = TPMT_PUBLIC;
; // class TPMT_PUBLIC
/**
 *  This structure is used when different symmetric block cipher (not XOR) algorithms may be selected. If the Object can be an ordinary parent (not a derivation parent), this must be the first field in the Object's parameter (see 12.2.3.7) field.
 */
var TPMT_SYM_DEF_OBJECT = (function (_super) {
    __extends(TPMT_SYM_DEF_OBJECT, _super);
    function TPMT_SYM_DEF_OBJECT(
        /**
         *  symmetric algorithm
         */
        algorithm, 
        /**
         *  key size in bits
         */
        keyBits, 
        /**
         *  encryption mode
         */
        mode) {
        if (algorithm === void 0) { algorithm = 0; }
        if (keyBits === void 0) { keyBits = 0; }
        if (mode === void 0) { mode = 0; }
        var _this = _super.call(this) || this;
        _this.algorithm = algorithm;
        _this.keyBits = keyBits;
        _this.mode = mode;
        return _this;
    }
    /** TpmMarshaller method */
    TPMT_SYM_DEF_OBJECT.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.nonStandardToTpm(this, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMT_SYM_DEF_OBJECT.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.nonStandardFromTpm(this, buf, pos);
        return pos;
    };
    return TPMT_SYM_DEF_OBJECT;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMT_SYM_DEF_OBJECT = TPMT_SYM_DEF_OBJECT;
; // class TPMT_SYM_DEF_OBJECT
/**
 *  The TPMT_SYM_DEF structure is used to select an algorithm to be used for parameter encryption in those cases when different symmetric algorithms may be selected.
 */
var TPMT_SYM_DEF = (function (_super) {
    __extends(TPMT_SYM_DEF, _super);
    function TPMT_SYM_DEF(
        /**
         *  symmetric algorithm
         */
        algorithm, 
        /**
         *  key size in bits
         */
        keyBits, 
        /**
         *  encryption mode
         */
        mode) {
        var _this = _super.call(this) || this;
        _this.algorithm = algorithm;
        _this.keyBits = keyBits;
        _this.mode = mode;
        return _this;
    }
    /** TpmMarshaller method */
    TPMT_SYM_DEF.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.nonStandardToTpm(this, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMT_SYM_DEF.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.nonStandardFromTpm(this, buf, pos);
        return pos;
    };
    return TPMT_SYM_DEF;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMT_SYM_DEF = TPMT_SYM_DEF;
; // class TPMT_SYM_DEF
/**
* This ticket is produced by TPM2_PolicySigned() and TPM2_PolicySecret() when the authorization has an expiration time. If nonceTPM was provided in the policy command, the ticket is computed by
*/
var TPMT_TK_AUTH = (function (_super) {
    __extends(TPMT_TK_AUTH, _super);
    function TPMT_TK_AUTH(
        /**
        * ticket structure tag
        */
        tag, 
        /**
        * the hierarchy of the object used to produce the ticket
        */
        hierarchy, 
        /**
        * This shall be the HMAC produced using a proof value of hierarchy.
        */
        digest) {
        if (tag === void 0) { tag = 0; }
        if (hierarchy === void 0) { hierarchy = null; }
        if (digest === void 0) { digest = null; }
        var _this = _super.call(this) || this;
        _this.tag = tag;
        _this.hierarchy = hierarchy;
        _this.digest = digest;
        return _this;
    }
    /** TpmMarshaller method */
    TPMT_TK_AUTH.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.tag, buf, 2, pos);
        pos = this.hierarchy.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.digest, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMT_TK_AUTH.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.tag = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.createFromTpm(TPM_HANDLE, buf, pos), this.hierarchy = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.digest = _c[0], pos = _c[1];
        return pos;
        var _a, _b, _c;
    };
    return TPMT_TK_AUTH;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMT_TK_AUTH = TPMT_TK_AUTH;
;
/**
* This ticket is produced by TPM2_Create() or TPM2_CreatePrimary(). It is used to bind the creation data to the object to which it applies. The ticket is computed by
*/
var TPMT_TK_CREATION = (function (_super) {
    __extends(TPMT_TK_CREATION, _super);
    function TPMT_TK_CREATION(
        /**
        * ticket structure tag
        */
        tag, 
        /**
        * the hierarchy containing name
        */
        hierarchy, 
        /**
        * This shall be the HMAC produced using a proof value of hierarchy.
        */
        digest) {
        if (tag === void 0) { tag = 0; }
        if (hierarchy === void 0) { hierarchy = null; }
        if (digest === void 0) { digest = null; }
        var _this = _super.call(this) || this;
        _this.tag = tag;
        _this.hierarchy = hierarchy;
        _this.digest = digest;
        return _this;
    }
    /** TpmMarshaller method */
    TPMT_TK_CREATION.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.tag, buf, 2, pos);
        pos = this.hierarchy.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.digest, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMT_TK_CREATION.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.tag = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.createFromTpm(TPM_HANDLE, buf, pos), this.hierarchy = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.digest = _c[0], pos = _c[1];
        return pos;
        var _a, _b, _c;
    };
    return TPMT_TK_CREATION;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMT_TK_CREATION = TPMT_TK_CREATION;
; // class TPMT_TK_CREATION
/**
* This ticket is produced by TPM2_SequenceComplete() when the message that was digested did not start with TPM_GENERATED_VALUE. The ticket is computed by
*/
var TPMT_TK_HASHCHECK = (function (_super) {
    __extends(TPMT_TK_HASHCHECK, _super);
    /**
    * This ticket is produced by TPM2_SequenceComplete() when the message that was digested did not start with TPM_GENERATED_VALUE. The ticket is computed by
    *
    * @param _tag ticket structure tag
    * @param _hierarchy the hierarchy
    * @param _digest This shall be the HMAC produced using a proof value of hierarchy.
    */
    function TPMT_TK_HASHCHECK(tag, hierarchy, digest) {
        if (tag === void 0) { tag = null; }
        if (hierarchy === void 0) { hierarchy = null; }
        if (digest === void 0) { digest = null; }
        var _this = _super.call(this) || this;
        _this.tag = tag;
        _this.hierarchy = hierarchy;
        _this.digest = digest;
        return _this;
    }
    /** TpmMarshaller method */
    TPMT_TK_HASHCHECK.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.tag, buf, 2, pos);
        pos = this.hierarchy.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.digest, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    TPMT_TK_HASHCHECK.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 2, pos), this.tag = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.createFromTpm(TPM_HANDLE, buf, pos), this.hierarchy = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.digest = _c[0], pos = _c[1];
        return pos;
        var _a, _b, _c;
    };
    return TPMT_TK_HASHCHECK;
}(TpmMarshaller_js_1.TpmStructure));
exports.TPMT_TK_HASHCHECK = TPMT_TK_HASHCHECK;
; // class TPMT_TK_HASHCHECK
/**
* This command is used to create a Primary Object under one of the Primary Seeds or a Temporary Object under TPM_RH_NULL. The command uses a TPM2B_PUBLIC as a template for the object to be created. The size of the unique field shall not be checked for consistency with the other object parameters. The command will create and load a Primary Object. The sensitive area is not returned.
*/
var CreatePrimaryResponse = (function (_super) {
    __extends(CreatePrimaryResponse, _super);
    /**
    * @param _handle handle of type TPM_HT_TRANSIENT for created Primary Object
    * @param _outPublic the public portion of the created object
    * @param _creationData contains a TPMT_CREATION_DATA
    * @param _creationHash digest of creationData using nameAlg of outPublic
    * @param _creationTicket ticket used by TPM2_CertifyCreation() to validate that the creation data was produced by the TPM
    * @param _name the name of the created object
    */
    function CreatePrimaryResponse(handle, outPublic, creationData, creationHash, creationTicket, name) {
        if (handle === void 0) { handle = null; }
        if (outPublic === void 0) { outPublic = null; }
        if (creationData === void 0) { creationData = null; }
        if (creationHash === void 0) { creationHash = null; }
        if (creationTicket === void 0) { creationTicket = null; }
        if (name === void 0) { name = null; }
        var _this = _super.call(this) || this;
        _this.handle = handle;
        _this.outPublic = outPublic;
        _this.creationData = creationData;
        _this.creationHash = creationHash;
        _this.creationTicket = creationTicket;
        _this.name = name;
        return _this;
    }
    /** TpmMarshaller method */
    CreatePrimaryResponse.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = this.handle.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.sizedToTpm(this.outPublic, buf, 2, pos);
        pos = this.creationData.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.creationHash, buf, pos);
        pos = this.creationTicket.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.name, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    CreatePrimaryResponse.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.createFromTpm(TPM_HANDLE, buf, pos), this.handle = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.sizedFromTpm(TPMT_PUBLIC, buf, 2, pos), this.outPublic = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.sizedFromTpm(TPMS_CREATION_DATA, buf, 2, pos), this.creationData = _c[0], pos = _c[1];
        _d = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.creationHash = _d[0], pos = _d[1];
        _e = TpmMarshaller_js_1.createFromTpm(TPMT_TK_CREATION, buf, pos), this.creationTicket = _e[0], pos = _e[1];
        _f = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.name = _f[0], pos = _f[1];
        return pos;
        var _a, _b, _c, _d, _e, _f;
    };
    return CreatePrimaryResponse;
}(TpmMarshaller_js_1.TpmStructure));
exports.CreatePrimaryResponse = CreatePrimaryResponse;
; // class CreatePrimaryResponse
/**
* This command returns various information regarding the TPM and its current state.
*/
var GetCapabilityResponse = (function (_super) {
    __extends(GetCapabilityResponse, _super);
    function GetCapabilityResponse(
        /**
         * flag to indicate if there are more values of this type
         */
        moreData, 
        /**
         * the capability data (One of TPML_ALG_PROPERTY, TPML_HANDLE, TPML_CCA, TPML_CC, TPML_CC, TPML_PCR_SELECTION, TPML_TAGGED_TPM_PROPERTY, TPML_TAGGED_PCR_PROPERTY, TPML_ECC_CURVE, TPML_TAGGED_POLICY)
         */
        capabilityData) {
        if (moreData === void 0) { moreData = null; }
        if (capabilityData === void 0) { capabilityData = null; }
        var _this = _super.call(this) || this;
        _this.moreData = moreData;
        _this.capabilityData = capabilityData;
        return _this;
    }
    /** TpmMarshaller method */
    GetCapabilityResponse.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm(this.moreData, buf, 1, pos);
        var capabilityDataCapability = 0;
        pos = TpmMarshaller_js_1.toTpm(this.capabilityData.GetUnionSelector(), buf, 4, pos);
        pos = this.capabilityData.toTpm(buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    GetCapabilityResponse.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm(buf, 1, pos), this.moreData = _a[0], pos = _a[1];
        var capabilityDataCapability = 0;
        _b = TpmMarshaller_js_1.fromTpm(buf, 4, pos), capabilityDataCapability = _b[0], pos = _b[1];
        this.capabilityData = createUnion('TPMU_CAPABILITIES', capabilityDataCapability);
        pos = this.capabilityData.fromTpm(buf, pos);
        return pos;
        var _a, _b;
    };
    return GetCapabilityResponse;
}(TpmMarshaller_js_1.TpmStructure));
exports.GetCapabilityResponse = GetCapabilityResponse;
;
/**
 *  This command allows access to the public area of a loaded object.
 */
var ReadPublicResponse = (function (_super) {
    __extends(ReadPublicResponse, _super);
    function ReadPublicResponse(
        /**
         *  structure containing the public area of an object
         */
        outPublic, 
        /**
         *  name of the object
         */
        name, 
        /**
         *  the Qualified Name of the object
         */
        qualifiedName) {
        if (outPublic === void 0) { outPublic = null; }
        if (name === void 0) { name = null; }
        if (qualifiedName === void 0) { qualifiedName = null; }
        var _this = _super.call(this) || this;
        _this.outPublic = outPublic;
        _this.name = name;
        _this.qualifiedName = qualifiedName;
        return _this;
    }
    /** TpmMarshaller method */
    ReadPublicResponse.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.sizedToTpm(this.outPublic, buf, 2, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.name, buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.qualifiedName, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    ReadPublicResponse.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.sizedFromTpm(TPMT_PUBLIC, buf, 2, pos), this.outPublic = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.name = _b[0], pos = _b[1];
        _c = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.qualifiedName = _c[0], pos = _c[1];
        return pos;
        var _a, _b, _c;
    };
    return ReadPublicResponse;
}(TpmMarshaller_js_1.TpmStructure));
exports.ReadPublicResponse = ReadPublicResponse;
; // class ReadPublicResponse
/**
* This command adds the last part of data, if any, to a hash/HMAC sequence and returns the result.
*/
var SequenceCompleteResponse = (function (_super) {
    __extends(SequenceCompleteResponse, _super);
    function SequenceCompleteResponse(
        /**
         * the returned HMAC or digest in a sized buffer
         */
        result, 
        /**
         * ticket indicating that the sequence of octets used to compute outDigest did not start with TPM_GENERATED_VALUE This is a NULL Ticket when the sequence is HMAC.
         */
        validation) {
        if (result === void 0) { result = null; }
        if (validation === void 0) { validation = null; }
        var _this = _super.call(this) || this;
        _this.result = result;
        _this.validation = validation;
        return _this;
    }
    /** TpmMarshaller method */
    SequenceCompleteResponse.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.result, buf, pos);
        pos = this.validation.toTpm(buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    SequenceCompleteResponse.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.result = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.createFromTpm(TPMT_TK_HASHCHECK, buf, pos), this.validation = _b[0], pos = _b[1];
        return pos;
        var _a, _b;
    };
    return SequenceCompleteResponse;
}(TpmMarshaller_js_1.TpmStructure));
exports.SequenceCompleteResponse = SequenceCompleteResponse;
; // class SequenceCompleteResponse
/**
* This command is used to start an authorization session using alternative methods of establishing the session key (sessionKey). The session key is then used to derive values used for authorization and for encrypting parameters.
*/
var StartAuthSessionResponse = (function (_super) {
    __extends(StartAuthSessionResponse, _super);
    function StartAuthSessionResponse(
        /**
        * handle for the newly created session
        */
        handle, 
        /**
        * the initial nonce from the TPM, used in the computation of the sessionKey
        */
        nonceTPM) {
        if (handle === void 0) { handle = null; }
        if (nonceTPM === void 0) { nonceTPM = null; }
        var _this = _super.call(this) || this;
        _this.handle = handle;
        _this.nonceTPM = nonceTPM;
        return _this;
    }
    /** TpmMarshaller method */
    StartAuthSessionResponse.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = this.handle.toTpm(buf, pos);
        pos = TpmMarshaller_js_1.toTpm2B(this.nonceTPM, buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    StartAuthSessionResponse.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.createFromTpm(TPM_HANDLE, buf, pos), this.handle = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.nonceTPM = _b[0], pos = _b[1];
        return pos;
        var _a, _b;
    };
    return StartAuthSessionResponse;
}(TpmMarshaller_js_1.TpmStructure));
exports.StartAuthSessionResponse = StartAuthSessionResponse;
; //class StartAuthSessionResponse
/**
* This command includes a secret-based authorization to a policy. The caller proves knowledge of the secret value using an authorization session using the authValue associated with authHandle. A password session, an HMAC session, or a policy session containing TPM2_PolicyAuthValue() or TPM2_PolicyPassword() will satisfy this requirement.
*/
var PolicySecretResponse = (function (_super) {
    __extends(PolicySecretResponse, _super);
    /**
    * This command includes a secret-based authorization to a policy. The caller proves knowledge of the secret value using an authorization session using the authValue associated with authHandle. A password session, an HMAC session, or a policy session containing TPM2_PolicyAuthValue() or TPM2_PolicyPassword() will satisfy this requirement.
    *
    * @param _timeout implementation-specific time value used to indicate to the TPM when the ticket expires
    * @param _policyTicket produced if the command succeeds and expiration in the command was non-zero ( See 23.2.5). This ticket will use the TPMT_ST_AUTH_SECRET structure tag
    */
    function PolicySecretResponse(timeout, policyTicket) {
        if (timeout === void 0) { timeout = null; }
        if (policyTicket === void 0) { policyTicket = null; }
        var _this = _super.call(this) || this;
        _this.timeout = timeout;
        _this.policyTicket = policyTicket;
        return _this;
    }
    /** TpmMarshaller method */
    PolicySecretResponse.prototype.toTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        pos = TpmMarshaller_js_1.toTpm2B(this.timeout, buf, pos);
        pos = this.policyTicket.toTpm(buf, pos);
        return pos;
    };
    /** TpmMarshaller method */
    PolicySecretResponse.prototype.fromTpm = function (buf, pos) {
        if (pos === void 0) { pos = 0; }
        _a = TpmMarshaller_js_1.fromTpm2B(buf, pos), this.timeout = _a[0], pos = _a[1];
        _b = TpmMarshaller_js_1.createFromTpm(TPMT_TK_AUTH, buf, pos), this.policyTicket = _b[0], pos = _b[1];
        return pos;
        var _a, _b;
    };
    return PolicySecretResponse;
}(TpmMarshaller_js_1.TpmStructure));
exports.PolicySecretResponse = PolicySecretResponse;
; // class PolicySecretResponse
