// azure-iot-security-tpm/lib/tpm_authentication_provider.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var dbg = require("debug");
var debug = dbg('azure-iot-device:TPMAuthenticationProvider');
var events_1 = require("events");
var machina = require("machina");
var azure_iot_common_1 = require("azure-iot-common");
/**
 * @private
 */
var TpmAuthenticationProvider = (function (_super) {
    __extends(TpmAuthenticationProvider, _super);
    /**
     * @private
     *
     * Initializes a new instance of the TpmAuthenticationProvider - users should only use the factory methods though.
     *
     * @param credentials       Credentials to be used by the device to connect to the IoT hub.
     * @param tpmSecurityClient That client that provides the signingFunction.
     */
    function TpmAuthenticationProvider(credentials, tpmSecurityClient) {
        var _this = _super.call(this) || this;
        _this.type = azure_iot_common_1.AuthenticationType.Token;
        _this.automaticRenewal = true;
        _this._tokenValidTimeInSeconds = 3600; // 1 hour
        _this._tokenRenewalMarginInSeconds = 900; // 15 minutes
        _this._credentials = credentials;
        _this._tpmSecurityClient = tpmSecurityClient;
        _this._fsm = new machina.Fsm({
            initialState: 'inactive',
            states: {
                inactive: {
                    _onEnter: function (err, callback) {
                        if (callback) {
                            callback(err);
                        }
                        else if (err) {
                            _this.emit('error', err);
                        }
                        if (_this._renewalTimeout) {
                            clearTimeout(_this._renewalTimeout);
                        }
                    },
                    activate: function (activateCallback) { return _this._fsm.transition('activating', activateCallback); },
                    getDeviceCredentials: function (callback) {
                        _this._fsm.handle('activate', function (err, result) {
                            if (err) {
                                callback(err);
                            }
                            else {
                                _this._fsm.handle('getDeviceCredentials', callback);
                            }
                        });
                    }
                },
                activating: {
                    _onEnter: function (callback, err) {
                        var newExpiry = Math.floor(Date.now() / 1000) + _this._tokenValidTimeInSeconds;
                        azure_iot_common_1.SharedAccessSignature.createWithSigningFunction(_this._credentials, newExpiry, _this._tpmSecurityClient.signWithIdentity.bind(_this._tpmSecurityClient), function (err, newSas) {
                            if (err) {
                                debug('Unable to create a new SAS token! - ' + err);
                                callback(err);
                                _this._fsm.transition('inactive');
                            }
                            else {
                                _this._credentials.sharedAccessSignature = newSas.toString();
                                _this._renewalTimeout = setTimeout(function () { return _this._renewToken(); }, (_this._tokenValidTimeInSeconds - _this._tokenRenewalMarginInSeconds) * 1000);
                                debug('Created a new sas token.');
                                _this._fsm.transition('active', callback);
                            }
                        });
                    },
                    '*': function () { return _this._fsm.deferUntilTransition(); }
                },
                active: {
                    _onEnter: function (callback) {
                        _this.emit('newTokenAvailable', _this._credentials);
                        callback(null, _this._credentials);
                    },
                    getDeviceCredentials: function (callback) {
                        callback(null, _this._credentials);
                    },
                    signingError: function (err) {
                        debug('Unable to create a new SAS token! - ' + err);
                        _this._fsm.transition('inactive', err);
                    },
                    signingSuccessful: function () {
                        debug('Created a new sas token.');
                        _this.emit('newTokenAvailable', _this._credentials);
                    }
                }
            }
        });
        return _this;
    }
    TpmAuthenticationProvider.prototype.getDeviceCredentials = function (callback) {
        this._fsm.handle('getDeviceCredentials', callback);
    };
    TpmAuthenticationProvider.prototype.updateSharedAccessSignature = function (sharedAccessSignature) {
        throw new azure_iot_common_1.errors.InvalidOperationError('cannot update a shared access signature when using TPM');
    };
    TpmAuthenticationProvider.prototype._renewToken = function () {
        var _this = this;
        if (this._renewalTimeout) {
            clearTimeout(this._renewalTimeout);
        }
        var newExpiry = Math.floor(Date.now() / 1000) + this._tokenValidTimeInSeconds;
        azure_iot_common_1.SharedAccessSignature.createWithSigningFunction(this._credentials, newExpiry, this._tpmSecurityClient.signWithIdentity.bind(this._tpmSecurityClient), function (err, newSas) {
            if (err) {
                _this._fsm.handle('signingError', err);
            }
            else {
                _this._credentials.sharedAccessSignature = newSas.toString();
                _this._renewalTimeout = setTimeout(function () { return _this._renewToken(); }, (_this._tokenValidTimeInSeconds - _this._tokenRenewalMarginInSeconds) * 1000);
                _this._fsm.handle('signingSuccessful');
            }
        });
    };
    /**
     * @method           module:azure-iot-security-tpm.TpmSecurityClient#fromTpmSecurityClient
     * @description      Returns an authentication provider.
     * @param {string}            deviceId          The device id for the client to be created.
     * @param {string}            iotHubHostname    The name of the IoT hub to be utilized.
     * @param {TpmSecurityClient} tpmSecurityClient The client which provides the tpm security interface. (Signing, activating, etc.)
     */
    TpmAuthenticationProvider.fromTpmSecurityClient = function (deviceId, iotHubHostname, tpmSecurityClient) {
        if (!deviceId) {
            throw new ReferenceError('deviceId cannot be \'' + deviceId + '\'');
        }
        if (!iotHubHostname) {
            throw new ReferenceError('iotHubHostname cannot be \'' + iotHubHostname + '\'');
        }
        if (!tpmSecurityClient) {
            throw new ReferenceError('tpmSecurityClient cannot be \'' + tpmSecurityClient + '\'');
        }
        var credentials = {
            host: iotHubHostname,
            deviceId: deviceId
        };
        return new TpmAuthenticationProvider(credentials, tpmSecurityClient);
    };
    return TpmAuthenticationProvider;
}(events_1.EventEmitter));
exports.TpmAuthenticationProvider = TpmAuthenticationProvider;
